{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f0e0702d",
   "metadata": {},
   "source": [
    "## Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc47c781",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "df = pd.read_parquet(\"/Users/isabell/Documents/projects/MAPKprot/full_proteome/result.parquet\")\n",
    "df_long = df[[\"Protein.Group\",\"Run\",\"PG.MaxLFQ\", \"Protein.Names\"]].drop_duplicates().reset_index(drop=True)\n",
    "df_long.columns = ['protein_id', 'filename', 'intensity', 'protein_name']\n",
    "df_long['filename'] = df_long['filename'].str.split('_').str[-1]\n",
    "df_long[\"filename\"] = pd.to_numeric(df_long[\"filename\"], errors=\"coerce\").astype(str)\n",
    "df_long"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1bc212a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read Excel file\n",
    "mapk_df = (\n",
    "    pd.read_excel(\"/Users/isabell/Documents/projects/MAPKprot/annotation.xlsx\")\n",
    "    .drop(columns=[\"Unnamed: 0\"], errors=\"ignore\")\n",
    "    .rename(columns=lambda x: x.strip().replace(\" \", \"_\"))\n",
    "    [[\"Sample_ID\", \"type\", \"alteration\",\"sex\",\n",
    "      \"age\", \"mapk\", \"location\"]]\n",
    ")\n",
    "mapk_df[\"filename\"] = mapk_df[\"Sample_ID\"].astype(str)\n",
    "mapk_df = mapk_df[[\"filename\",\"type\", \"alteration\",\"sex\",\n",
    "      \"age\", \"mapk\", \"location\"]]\n",
    "mapk_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4204ef5f",
   "metadata": {},
   "outputs": [],
   "source": [
    "mapk_df.type.unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2ef76f28",
   "metadata": {},
   "source": [
    "## Load CoPro"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "296f10b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import random\n",
    "from pathlib import Path\n",
    "import warnings\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scipy as sp\n",
    "import anndata as ad\n",
    "import scanpy as sc\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib as mpl\n",
    "\n",
    "import copro as cp\n",
    "\n",
    "random.seed(42)\n",
    "\n",
    "cwd = Path('.').resolve()\n",
    "root = cwd.parent\n",
    "os.chdir(root)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f1168949",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = cp.read.proteins_long_from_df(intensities_df=df_long, filename_annotation_df=mapk_df) #, protein_annotation_df=Karayel_2020_meta_var)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fffeeb92",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9114e602",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = adata[~adata.obs[\"type\"].isna(), :].copy()\n",
    "adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2bd96931",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.var.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fcecba50",
   "metadata": {},
   "outputs": [],
   "source": [
    "set(adata.obs.type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "678b101f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Colors for plotting\n",
    "adata.uns['colors_type'] = {\n",
    "    #\"DMG\": \"#377EB8\",       # blue\n",
    "    #\"DMG_MAPK\": \"#109ABC\",\n",
    "    \"DNT\": \"#C850D3\",       # purple\n",
    "    \"EVNCYT\": \"#3A1C0B\",    # brown\n",
    "    \"GG\": \"#F781BF\",        # pink\n",
    "    \"PA_CORT\": \"#66C2A5\",   # teal\n",
    "    \"PA_INF\": \"#117E5A\",  # dark turquoise\n",
    "    \"PA_MID\": \"#1B759E\",  # dark turquoise\n",
    "    \"PA_SPINE\":\"#184D7E\",  # dark turquoise \n",
    "    \"PXA\": \"#471F7F\",\n",
    "    \"CN\": \"#E41A1C\",        # red\n",
    "    \"MNG\": \"#999999\",       # grey\n",
    "    \"EPN_PFA\": \"#FF7F00\",   # orange\n",
    "    \"EPN_PFB\": \"#FFAA00\",   \n",
    " }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f33d3172",
   "metadata": {},
   "outputs": [],
   "source": [
    "cp.pl.n_samples_by_category(\n",
    "        adata, \n",
    "        category_cols='type',\n",
    "        color_scheme=adata.uns['colors_type'],\n",
    "        sort_by_counts=False\n",
    "        )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c481f29",
   "metadata": {},
   "outputs": [],
   "source": [
    "cp.pl.n_detected_proteins_per_sample(\n",
    "    adata,\n",
    "    zero_to_na=True,\n",
    "    group_by='type',\n",
    "    group_by_order=adata.uns['colors_type'].keys(),\n",
    "    color_scheme=adata.uns['colors_type'],\n",
    "    xlabel_rotation=90,\n",
    "    group_by_label_rotation=40,\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bbe1ac8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "cp.pp.filter_obs_by_min_nr_var(adata, min_nr=6000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "06653d26",
   "metadata": {},
   "outputs": [],
   "source": [
    "cp.pl.n_detected_proteins_per_sample(\n",
    "    adata,\n",
    "    zero_to_na=True,\n",
    "    group_by='type',\n",
    "    group_by_order=adata.uns['colors_type'].keys(),\n",
    "    color_scheme=adata.uns['colors_type'],\n",
    "    xlabel_rotation=90,\n",
    "    group_by_label_rotation=40,\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "07265f87",
   "metadata": {},
   "outputs": [],
   "source": [
    "cp.pl.var_completeness(adata, zero_to_na=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a5d7dde",
   "metadata": {},
   "outputs": [],
   "source": [
    "cp.pp.filter_var_completeness(adata, min_fraction=1, group_by='type')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "213c02c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "cp.pl.var_completeness(adata, zero_to_na=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7cd9cf63",
   "metadata": {},
   "outputs": [],
   "source": [
    "cp.pp.filter_var_completeness(adata, min_fraction=0.33, zero_to_na=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "329ce033",
   "metadata": {},
   "outputs": [],
   "source": [
    "cp.pl.var_completeness(adata, zero_to_na=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5739f112",
   "metadata": {},
   "outputs": [],
   "source": [
    "cp.pl.n_detected_proteins_per_sample(\n",
    "    adata,\n",
    "    zero_to_na=True,\n",
    "    group_by='type',\n",
    "    group_by_order=adata.uns['colors_type'].keys(),\n",
    "    color_scheme=adata.uns['colors_type'],\n",
    "    xlabel_rotation=90,\n",
    "    group_by_label_rotation=40,\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e4badf4a",
   "metadata": {},
   "outputs": [],
   "source": [
    "cp.pp.calculate_groupwise_cv(\n",
    "    adata,\n",
    "    groupby='type'\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c2b9ca3",
   "metadata": {},
   "outputs": [],
   "source": [
    "cp.pl.cv_distribution(\n",
    "    adata,\n",
    "    figsize=(6, 4),\n",
    "    group_label_rotation=55,\n",
    "    order=adata.uns['colors_type'].keys(),\n",
    "    color_scheme=adata.uns['colors_type'],\n",
    "    hline=0.5\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29414a73",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.X = np.log2(adata.X)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c155e217",
   "metadata": {},
   "outputs": [],
   "source": [
    "cp.pl.intensity_distribution_per_obs(\n",
    "    adata,\n",
    "    group_by='type',\n",
    "    zero_to_na=True,\n",
    "    group_by_order=adata.uns['colors_type'].keys(),\n",
    "    color_scheme=adata.uns['colors_type'],\n",
    "    group_by_label_rotation=45,\n",
    "    xlabel_rotation=55,\n",
    "    show=True,\n",
    "    figsize=(9, 6)\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "89ca28df",
   "metadata": {},
   "outputs": [],
   "source": [
    "cp.pp.median_normalize(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b34cc628",
   "metadata": {},
   "outputs": [],
   "source": [
    "cp.pl.intensity_distribution_per_obs(\n",
    "    adata,\n",
    "    group_by='type',\n",
    "    zero_to_na=True,\n",
    "    group_by_order=adata.uns['colors_type'].keys(),\n",
    "    color_scheme=adata.uns['colors_type'],\n",
    "    group_by_label_rotation=45,\n",
    "    xlabel_rotation=55,\n",
    "    show=True,\n",
    "    figsize=(9, 6)\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9947c0fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "cp.pp.impute_downshift(adata, width=0.3, downshift=1.8, random_state=123)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "312ac421",
   "metadata": {},
   "outputs": [],
   "source": [
    "measured_n = int((adata.layers[\"bool_imputed\"] == 0).sum())\n",
    "imputed_n  = int((adata.layers[\"bool_imputed\"] == 1).sum())\n",
    "print(\"Measured count:\", measured_n, \" | Imputed count:\", imputed_n)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "06f9ac15",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1) Single combined histogram\n",
    "cp.pl.intensity_hist_imputed(adata, density=False)\n",
    "\n",
    "# 2) Per-sample small multiples for all samples\n",
    "#cp.pl.intensity_hist_imputed(adata, per_sample=True, ncols=5, legend_loc=\"upper right\", density=False, figsize=(17, 12))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7810f73a",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.pca(adata)\n",
    "sc.pl.pca_variance_ratio(adata, n_pcs=50, log=True)\n",
    "sc.pl.pca(\n",
    "    adata,\n",
    "    color='type',\n",
    "    dimensions=(0, 1),\n",
    "    ncols=2,\n",
    "    size=90,\n",
    "    palette=adata.uns['colors_type'],\n",
    "    )\n",
    "sc.pl.pca(\n",
    "    adata,\n",
    "    color='alteration',\n",
    "    dimensions=(0, 1),\n",
    "    ncols=2,\n",
    "    size=90,\n",
    "    #palette=adata.uns['colors_cell_type'],\n",
    "    )\n",
    "sc.pl.pca(\n",
    "    adata,\n",
    "    color='sex',\n",
    "    dimensions=(0, 1),\n",
    "    ncols=2,\n",
    "    size=90,\n",
    "    #palette=adata.uns['colors_cell_type'],\n",
    "    )\n",
    "sc.pl.pca(\n",
    "    adata,\n",
    "    color='age',\n",
    "    dimensions=(0, 1),\n",
    "    ncols=2,\n",
    "    size=90,\n",
    "    #palette=adata.uns['colors_cell_type'],\n",
    "    )\n",
    "sc.pl.pca(\n",
    "    adata,\n",
    "    color='mapk',\n",
    "    dimensions=(0, 1),\n",
    "    ncols=2,\n",
    "    size=90,\n",
    "    #palette=adata.uns['colors_cell_type'],\n",
    "    )\n",
    "sc.pl.pca(\n",
    "    adata,\n",
    "    color='location',\n",
    "    dimensions=(0, 1),\n",
    "    ncols=2,\n",
    "    size=90,\n",
    "    #palette=adata.uns['colors_cell_type'],\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c439e33c",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.neighbors(adata, n_neighbors=5)\n",
    "sc.tl.umap(adata)\n",
    "sc.pl.umap(\n",
    "    adata,\n",
    "    color='type',\n",
    "    size=100,\n",
    "    palette=adata.uns['colors_type'],\n",
    "    )\n",
    "sc.pl.umap(\n",
    "    adata,\n",
    "    color='alteration',\n",
    "    size=100,\n",
    "    #palette=adata.uns['colors_cell_type'],\n",
    "    )\n",
    "sc.pl.umap(\n",
    "    adata,\n",
    "    color='sex',\n",
    "    size=100,\n",
    "    #palette=adata.uns['colors_cell_type'],\n",
    "    )\n",
    "sc.pl.umap(\n",
    "    adata,\n",
    "    color='age',\n",
    "    size=100,\n",
    "    #palette=adata.uns['colors_cell_type'],\n",
    "    )\n",
    "sc.pl.umap(\n",
    "    adata,\n",
    "    color='mapk',\n",
    "    size=100,\n",
    "    #palette=adata.uns['colors_cell_type'],\n",
    "    )\n",
    "sc.pl.umap(\n",
    "    adata,\n",
    "    color='location',\n",
    "    size=100,\n",
    "    #palette=adata.uns['colors_cell_type'],\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a17228f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "#sc.pp.combat(\n",
    "#    adata, \n",
    "#    key='batch_id', \n",
    "#    #covariates='tumor_type', \n",
    "#    inplace=True\n",
    "#    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d69c4cff",
   "metadata": {},
   "outputs": [],
   "source": [
    "cp.pl.obs_correlation_matrix(\n",
    "    adata,\n",
    "    method=\"pearson\",\n",
    "    groupby=\"type\",\n",
    "    color_scheme=adata.uns[\"colors_type\"],\n",
    "    xticklabels=True,\n",
    "    figsize=(13, 9),\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "726cb0c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "cp.pl.obs_correlation_matrix(\n",
    "    adata,\n",
    "    method=\"pearson\",\n",
    "    groupby=\"mapk\",\n",
    "    #color_scheme=adata.uns[\"colors_type\"],\n",
    "    xticklabels=True,\n",
    "    figsize=(13, 9),\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "58830a54",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Mark the 250 most variable proteins (HVG-style)\n",
    "sc.pp.highly_variable_genes(\n",
    "    adata,\n",
    "    n_top_genes=250,\n",
    "    inplace=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d7cc3ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get those variable features\n",
    "adata_hvg = adata[:, adata.var['highly_variable']].copy()\n",
    "\n",
    "# Clustered heatmap (Scanpy style)\n",
    "sc.pl.clustermap(\n",
    "    adata_hvg,\n",
    "    obs_keys='mapk',\n",
    "    z_score=1,\n",
    "    figsize=(8,6),\n",
    "    dendrogram_ratio=0.15,\n",
    "    show=True,\n",
    "    xticklabels=False\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d3677388",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_hvg.var.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e38e0b8",
   "metadata": {},
   "source": [
    "## What's missing:\n",
    "\n",
    "- (optional) log10 intensity vs abundance rank plot with option to highlight specific proteins > fig EV1D\n",
    "- (optional) clustering of selected vars (e.g. most variable, diff. exp.) >> fig 2B\n",
    "- (optional) plotting 'profiles' for each 'cluster' >> fig 2B\n",
    "- differential analysis (t-test)\n",
    "- (optional) differential analysis (anova)\n",
    "- volcano plot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "319e72bc",
   "metadata": {},
   "source": [
    "## Differential analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e64555a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "from scipy import stats\n",
    "\n",
    "def ttest_ind_all_vars(\n",
    "    adata,\n",
    "    group1_query: str,\n",
    "    group2_query: str,\n",
    "    equal_var: bool = True,\n",
    "):\n",
    "    \"\"\"\n",
    "    Perform scipy.stats.ttest_ind for all variables in adata,\n",
    "    compute log2 fold changes, and BH-FDR — with strict NA checking.\n",
    "\n",
    "    Raises:\n",
    "        ValueError if any NA appears in either group's expression matrix.\n",
    "\n",
    "    Returns:\n",
    "        DataFrame with mean1, mean2, log2FC, t_stat, p_val, p_adj\n",
    "    \"\"\"\n",
    "\n",
    "    # select obs indices\n",
    "    idx1 = adata.obs.query(group1_query).index\n",
    "    idx2 = adata.obs.query(group2_query).index\n",
    "\n",
    "    X = adata.X\n",
    "\n",
    "    # slice rows for each group\n",
    "    X1 = X[adata.obs.index.isin(idx1), :]\n",
    "    X2 = X[adata.obs.index.isin(idx2), :]\n",
    "\n",
    "    # convert to dense if sparse\n",
    "    if not isinstance(X1, np.ndarray):\n",
    "        X1 = X1.toarray()\n",
    "        X2 = X2.toarray()\n",
    "\n",
    "    # strict NA validation\n",
    "    if np.isnan(X1).any() or np.isnan(X2).any():\n",
    "        raise ValueError(\"Input expression matrix contains NA values — clean your data first.\")\n",
    "\n",
    "    # compute means\n",
    "    mean1 = X1.mean(axis=0)\n",
    "    mean2 = X2.mean(axis=0)\n",
    "\n",
    "    # check for zeros that would break log2 fold-change\n",
    "    if np.any(mean1 == 0) or np.any(mean2 == 0):\n",
    "        raise ValueError(\"Zero mean detected in one of the groups; log2 fold change undefined.\")\n",
    "\n",
    "    # log2 fold change\n",
    "    log2fc = np.log2(mean1 / mean2)\n",
    "\n",
    "    # t-test\n",
    "    t_stats, p_vals = stats.ttest_ind(\n",
    "        X1, X2,\n",
    "        axis=0,\n",
    "        equal_var=equal_var\n",
    "    )\n",
    "\n",
    "    # BH-FDR (Benjamini–Hochberg)\n",
    "    p = p_vals.copy()\n",
    "    n = p.size\n",
    "\n",
    "    order = np.argsort(p)\n",
    "    ranks = np.arange(1, n + 1, dtype=float)\n",
    "\n",
    "    bh = (p[order] * n) / ranks\n",
    "    bh = np.minimum.accumulate(bh[::-1])[::-1]\n",
    "\n",
    "    p_adj = np.empty_like(bh)\n",
    "    p_adj[order] = bh\n",
    "\n",
    "    return pd.DataFrame({\n",
    "        \"mean1\": mean1,\n",
    "        \"mean2\": mean2,\n",
    "        \"log2FC\": log2fc,\n",
    "        \"t_stat\": t_stats,\n",
    "        \"p_val\": p_vals,\n",
    "        \"p_adj\": p_adj,\n",
    "    }, index=adata.var_names)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b3d07268",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "def volcano_plot(\n",
    "    res_df,\n",
    "    fc_col: str = \"log2FC\",\n",
    "    p_col: str = \"p_adj\",   # or \"p_val\" if you prefer\n",
    "    fc_thresh: float = 0.3,\n",
    "    p_thresh: float = 0.01,\n",
    "    top_labels: int = 10,\n",
    "    label_col: str | None = None,\n",
    "    title: str | None = None,\n",
    "    ax=None,\n",
    "):\n",
    "    \"\"\"\n",
    "    Simple volcano plot from ttest_ind_all_vars results.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    res_df : pandas.DataFrame\n",
    "        Result of ttest_ind_all_vars (index = features, columns incl. log2FC, p_adj/p_val).\n",
    "    fc_col : str\n",
    "        Column name for log2 fold change.\n",
    "    p_col : str\n",
    "        Column name for p-values (e.g. 'p_adj' or 'p_val').\n",
    "    fc_thresh : float\n",
    "        Absolute log2FC threshold for calling points 'significant'.\n",
    "    p_thresh : float\n",
    "        P-value threshold for calling points 'significant'.\n",
    "    top_labels : int\n",
    "        Number of most significant points to label (by p-value).\n",
    "    label_col : str or None\n",
    "        Column to use for labels. If None, use res_df.index.\n",
    "    title : str or None\n",
    "        Plot title.\n",
    "    ax : matplotlib Axes, optional\n",
    "        Axes to plot into. If None, a new figure and axes are created.\n",
    "    \"\"\"\n",
    "\n",
    "    if ax is None:\n",
    "        fig, ax = plt.subplots(figsize=(6, 5))\n",
    "\n",
    "    # extract values\n",
    "    log2fc = res_df[fc_col].values\n",
    "    pvals = res_df[p_col].values\n",
    "\n",
    "    # -log10 p\n",
    "    neg_log10_p = -np.log10(pvals)\n",
    "\n",
    "    # significance masks\n",
    "    sig_mask = (np.abs(log2fc) >= fc_thresh) & (pvals <= p_thresh)\n",
    "    up_mask = sig_mask & (log2fc > 0)\n",
    "    down_mask = sig_mask & (log2fc < 0)\n",
    "    non_sig_mask = ~sig_mask\n",
    "\n",
    "    # scatter plot\n",
    "    ax.scatter(\n",
    "        log2fc[non_sig_mask],\n",
    "        neg_log10_p[non_sig_mask],\n",
    "        s=10,\n",
    "        alpha=0.5,\n",
    "        label=\"n.s.\",\n",
    "    )\n",
    "    ax.scatter(\n",
    "        log2fc[up_mask],\n",
    "        neg_log10_p[up_mask],\n",
    "        s=12,\n",
    "        alpha=0.8,\n",
    "        label=\"up\",\n",
    "    )\n",
    "    ax.scatter(\n",
    "        log2fc[down_mask],\n",
    "        neg_log10_p[down_mask],\n",
    "        s=12,\n",
    "        alpha=0.8,\n",
    "        label=\"down\",\n",
    "    )\n",
    "\n",
    "    # threshold lines\n",
    "    ax.axvline(fc_thresh, color=\"grey\", linestyle=\"--\", linewidth=1)\n",
    "    ax.axvline(-fc_thresh, color=\"grey\", linestyle=\"--\", linewidth=1)\n",
    "    ax.axhline(-np.log10(p_thresh), color=\"grey\", linestyle=\"--\", linewidth=1)\n",
    "\n",
    "    # axis labels\n",
    "    ax.set_xlabel(\"log2 fold change\")\n",
    "    ax.set_ylabel(f\"-log10({p_col})\")\n",
    "\n",
    "    if title is None:\n",
    "        title = f\"Volcano plot ({p_col}, |log2FC| ≥ {fc_thresh}, p ≤ {p_thresh})\"\n",
    "    ax.set_title(title)\n",
    "\n",
    "    # label top hits\n",
    "    if top_labels > 0:\n",
    "        # sort by p-value\n",
    "        order = np.argsort(pvals)\n",
    "        top_idx = order[:top_labels]\n",
    "\n",
    "        if label_col is None:\n",
    "            labels = res_df.index.to_numpy()\n",
    "        else:\n",
    "            labels = res_df[label_col].to_numpy()\n",
    "\n",
    "        for i in top_idx:\n",
    "            ax.text(\n",
    "                log2fc[i],\n",
    "                neg_log10_p[i],\n",
    "                str(labels[i]),\n",
    "                fontsize=8,\n",
    "                ha=\"center\",\n",
    "                va=\"bottom\",\n",
    "            )\n",
    "\n",
    "    ax.legend(frameon=False)\n",
    "    ax.spines[\"top\"].set_visible(False)\n",
    "    ax.spines[\"right\"].set_visible(False)\n",
    "\n",
    "    return ax\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fbce163d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import statsmodels.formula.api as smf\n",
    "\n",
    "def group_difference_with_covariates(\n",
    "    adata,\n",
    "    group: str,\n",
    "    covariates: list,\n",
    "    strict_na: bool = True,\n",
    "):\n",
    "    \"\"\"\n",
    "    Test whether groups differ after adjusting for covariates using\n",
    "    statsmodels OLS for all variables in adata.X.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    adata : AnnData\n",
    "        Expression matrix (n_obs × n_vars)\n",
    "    group : str\n",
    "        Column name in adata.obs defining the main group (tested effect)\n",
    "    covariates : list[str]\n",
    "        Additional covariates to adjust for (categorical or continuous)\n",
    "    strict_na : bool\n",
    "        If True, error on NA values.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    DataFrame:\n",
    "        index = var_names\n",
    "        columns = ['coef', 'p_val', 'p_adj']\n",
    "    \"\"\"\n",
    "\n",
    "    # Build data frame with covariates\n",
    "    meta = adata.obs[[group] + covariates].copy()\n",
    "\n",
    "    # handle categorical variables\n",
    "    for col in meta:\n",
    "        if meta[col].dtype.name == \"category\":\n",
    "            meta[col] = meta[col].astype(str)\n",
    "\n",
    "    X = adata.X\n",
    "    if not isinstance(X, np.ndarray):\n",
    "        X = X.toarray()\n",
    "\n",
    "    if strict_na and np.isnan(X).any():\n",
    "        raise ValueError(\"Expression matrix contains NA values.\")\n",
    "\n",
    "    results = []\n",
    "    n_vars = X.shape[1]\n",
    "\n",
    "    # Build formula string\n",
    "    rhs = \" + \".join(\n",
    "        [f\"C({group})\"] +\n",
    "        [f\"C({c})\" if meta[c].dtype == object else c for c in covariates]\n",
    "    )\n",
    "\n",
    "    # Loop over variables\n",
    "    for idx, var_name in enumerate(adata.var_names):\n",
    "        meta_local = meta.copy()\n",
    "        meta_local[\"y\"] = X[:, idx]\n",
    "\n",
    "        if strict_na and np.isnan(meta_local[\"y\"]).any():\n",
    "            raise ValueError(f\"NA values found in variable {var_name}\")\n",
    "\n",
    "        formula = f\"y ~ {rhs}\"\n",
    "        model = smf.ols(formula, data=meta_local).fit()\n",
    "\n",
    "        # extract the main group comparisons\n",
    "        # (assuming two groups; extension possible)\n",
    "        terms = model.params.index\n",
    "        group_terms = [t for t in terms if t.startswith(f\"C({group})\")]\n",
    "\n",
    "        if len(group_terms) == 0:\n",
    "            raise ValueError(\"Group variable must have at least 2 levels.\")\n",
    "\n",
    "        # For a binary group: single coefficient\n",
    "        term = group_terms[0]\n",
    "        coef = model.params[term]\n",
    "        pval = model.pvalues[term]\n",
    "\n",
    "        results.append((var_name, coef, pval))\n",
    "\n",
    "    df = pd.DataFrame(results, columns=[\"var\", \"coef\", \"p_val\"]).set_index(\"var\")\n",
    "\n",
    "    # BH correction\n",
    "    p = df[\"p_val\"].values\n",
    "    order = np.argsort(p)\n",
    "    ranks = np.arange(1, len(p) + 1)\n",
    "    bh = (p[order] * len(p)) / ranks\n",
    "    bh = np.minimum.accumulate(bh[::-1])[::-1]\n",
    "    df[\"p_adj\"] = np.nan\n",
    "    df.iloc[order, df.columns.get_loc(\"p_adj\")] = bh\n",
    "\n",
    "    return df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3dea2ae8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "def volcano_plot(res, covar_threshold=0.05, p_adj_threshold=0.05, log2fc_threshold=0.3):\n",
    "    # Copy to avoid modifying original\n",
    "    df = res.copy()\n",
    "\n",
    "    # Avoid log10(0) → inf\n",
    "    # Replace zeros (if any) by smallest non-zero p-value or a tiny number\n",
    "    tiny = 1e-300\n",
    "    p = df[\"p_adj\"].replace(0, np.nan)\n",
    "    min_nonzero = p[p > 0].min()\n",
    "    df[\"p_adj_safe\"] = df[\"p_adj\"].replace(0, min_nonzero if not np.isnan(min_nonzero) else tiny)\n",
    "\n",
    "    # Compute -log10(p_adj)\n",
    "    df[\"neg_log10_p\"] = -np.log10(df[\"p_adj_safe\"])\n",
    "\n",
    "    # Boolean mask for covariate-significant points\n",
    "    sig = df[\"covar_p_adj\"] <= covar_threshold\n",
    "\n",
    "    plt.figure(figsize=(6, 5))\n",
    "\n",
    "    # Non-significant points\n",
    "    plt.scatter(\n",
    "        df.loc[~sig, \"log2FC\"],\n",
    "        df.loc[~sig, \"neg_log10_p\"],\n",
    "        s=10,\n",
    "        color=\"lightgray\",\n",
    "        alpha=0.7,\n",
    "        label=f\"covar_p_adj > {covar_threshold}\"\n",
    "    )\n",
    "\n",
    "    # Significant points\n",
    "    plt.scatter(\n",
    "        df.loc[sig, \"log2FC\"],\n",
    "        df.loc[sig, \"neg_log10_p\"],\n",
    "        s=10,\n",
    "        color=\"crimson\",\n",
    "        alpha=0.8,\n",
    "        label=f\"covar_p_adj ≤ {covar_threshold}\"\n",
    "    )\n",
    "\n",
    "    plt.axhline(-np.log10(p_adj_threshold), linestyle=\"--\", color=\"black\", linewidth=1, alpha=0.5)\n",
    "    plt.axvline(-log2fc_threshold, linestyle=\"--\", color=\"black\", linewidth=1, alpha=0.5)\n",
    "    plt.axvline(log2fc_threshold, linestyle=\"--\", color=\"black\", linewidth=1, alpha=0.5)\n",
    "\n",
    "    plt.xlabel(\"log2 fold change\")\n",
    "    plt.ylabel(\"-log10(p_adj)\")\n",
    "    plt.title(\"Volcano plot\")\n",
    "    plt.legend(frameon=False)\n",
    "    plt.tight_layout()\n",
    "    plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b782bf3",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_covariates = group_difference_with_covariates(\n",
    "    adata,\n",
    "    group=\"sex\",\n",
    "    covariates=[\"type\",\"age\"],\n",
    ")\n",
    "res_ttest = ttest_ind_all_vars(\n",
    "    adata,\n",
    "    group1_query='sex == \"Female\"',\n",
    "    group2_query='sex == \"Male\"'\n",
    ")\n",
    "\n",
    "res_covariates = res_covariates.add_prefix(\"covar_\")\n",
    "res_merged = res_ttest.join(res_covariates, how=\"inner\")\n",
    "res_merged.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ee5d9301",
   "metadata": {},
   "outputs": [],
   "source": [
    "volcano_plot(res_merged, covar_threshold=0.01)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "98fff61d",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_covariates = group_difference_with_covariates(\n",
    "    adata,\n",
    "    group=\"mapk\",\n",
    "    covariates=[\"age\",\"sex\"], # \"age\", \"type\"\n",
    ")\n",
    "res_ttest = ttest_ind_all_vars(\n",
    "    adata,\n",
    "    group1_query='mapk == \"positive\"',\n",
    "    group2_query='mapk == \"negative\"'\n",
    ")\n",
    "\n",
    "res_covariates = res_covariates.add_prefix(\"covar_\")\n",
    "res_merged = res_ttest.join(res_covariates, how=\"inner\")\n",
    "res_merged.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "52736ffd",
   "metadata": {},
   "outputs": [],
   "source": [
    "mapk_all = res_merged.index.tolist()\n",
    "with open(\"mapk_all.txt\", \"w\") as f:\n",
    "    for item in mapk_all:\n",
    "        f.write(str(item) + \"\\n\")\n",
    "\n",
    "mapk_up = res_merged[(res_merged[\"p_adj\"] < 0.05) & (res_merged[\"log2FC\"] > 0.3)].index.tolist()\n",
    "with open(\"mapk_up.txt\", \"w\") as f:\n",
    "    for item in mapk_up:\n",
    "        f.write(str(item) + \"\\n\")\n",
    "\n",
    "mapk_down = res_merged[(res_merged[\"p_adj\"] < 0.05) & (res_merged[\"log2FC\"] < -0.3)].index.tolist()\n",
    "with open(\"mapk_down.txt\", \"w\") as f:\n",
    "    for item in mapk_down:\n",
    "        f.write(str(item) + \"\\n\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ce6fe7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "volcano_plot(res_merged, covar_threshold=0.05)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ffb7d036",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12ee8d70",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_sub = adata[adata.obs[\"type\"].isin([\"PA_SPINE\", \"PA_INF\",\"PA_CORT\", \"PA_MID\"]), :].copy()\n",
    "\n",
    "res_covariates = group_difference_with_covariates(\n",
    "    adata_sub,\n",
    "    group=\"location\",\n",
    "    covariates=[\"sex\",\"age\"],\n",
    ")\n",
    "res_ttest = ttest_ind_all_vars(\n",
    "    adata_sub,\n",
    "    group1_query='location == \"Spinal\"',\n",
    "    group2_query='location == \"Posterior fossa\"'\n",
    ")\n",
    "\n",
    "res_covariates = res_covariates.add_prefix(\"covar_\")\n",
    "res_merged = res_ttest.join(res_covariates, how=\"inner\")\n",
    "res_merged.sort_values(\"p_adj\").head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "841ac7a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "volcano_plot(res_merged, covar_threshold=0.05)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5ee091d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_sub = adata[adata.obs[\"alteration\"].isin([\"BRAF\",\"KIAA1549::BRAF\"]), :].copy()\n",
    "\n",
    "res_covariates = group_difference_with_covariates(\n",
    "    adata_sub,\n",
    "    group=\"alteration\",\n",
    "    covariates=[\"sex\",\"age\"], #,\"type\"],\n",
    ")\n",
    "res_ttest = ttest_ind_all_vars(\n",
    "    adata_sub,\n",
    "    group1_query='alteration == \"BRAF\"',\n",
    "    group2_query='alteration == \"KIAA1549::BRAF\"'\n",
    ")\n",
    "\n",
    "res_covariates = res_covariates.add_prefix(\"covar_\")\n",
    "res_merged = res_ttest.join(res_covariates, how=\"inner\")\n",
    "res_merged.sort_values(\"p_adj\").head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e315cd9",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_merged[(res_merged[\"p_adj\"] < 0.05) & (np.abs(res_merged[\"log2FC\"]) > 0.3)].sort_values(\"covar_p_adj\").head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "be86f8c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "volcano_plot(res_merged, covar_threshold=0.01)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bd2f65e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "braf_up = res_merged[(res_merged[\"p_adj\"] < 0.05) & (res_merged[\"log2FC\"] > 0.3)].index.tolist()\n",
    "with open(\"braf_up.txt\", \"w\") as f:\n",
    "    for item in braf_up:\n",
    "        f.write(str(item) + \"\\n\")\n",
    "\n",
    "braf_down = res_merged[(res_merged[\"p_adj\"] < 0.05) & (res_merged[\"log2FC\"] < -0.3)].index.tolist()\n",
    "with open(\"braf_down.txt\", \"w\") as f:\n",
    "    for item in braf_down:\n",
    "        f.write(str(item) + \"\\n\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c32a0474",
   "metadata": {},
   "outputs": [],
   "source": [
    "#! pip install gseapy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "abb0602f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gseapy as gp\n",
    "\n",
    "# Download KEGG gene sets (one-time)\n",
    "lib = gp.get_library(name=\"KEGG_2016\", organism=\"Human\")\n",
    "\n",
    "# Pick the MAPK pathway\n",
    "# (exact key name may vary slightly; print(list(lib.keys())[:20]) to inspect)\n",
    "mapk_key = \"MAPK signaling pathway Homo sapiens hsa04010\"\n",
    "mapk_genes = lib[mapk_key]\n",
    "\n",
    "# Wrap in dict for ssGSEA\n",
    "gene_sets = {mapk_key: mapk_genes}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6db86c8d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import mygene\n",
    "import pandas as pd\n",
    "\n",
    "mg = mygene.MyGeneInfo()\n",
    "\n",
    "def symbols_to_uniprot(gene_symbols, species=\"human\"):\n",
    "    \"\"\"\n",
    "    Convert a list of gene symbols to UniProt accessions.\n",
    "    Returns a list of UniProt IDs (deduplicated).\n",
    "    \"\"\"\n",
    "    res = mg.querymany(\n",
    "        gene_symbols,\n",
    "        scopes=\"symbol\",\n",
    "        fields=\"uniprot.Swiss-Prot,uniprot.TrEMBL\",\n",
    "        species=species\n",
    "    )\n",
    "    df = pd.DataFrame(res)\n",
    "\n",
    "    # Drop entries that weren't mapped\n",
    "    if \"notfound\" in df.columns:\n",
    "        df = df[~df[\"notfound\"].fillna(False)]\n",
    "\n",
    "    # ---- Normalize 'uniprot' field into a flat list of strings ----\n",
    "    def normalize_uniprot(u):\n",
    "        # u can be: str, dict, list, or NaN\n",
    "        if isinstance(u, str):\n",
    "            return [u]\n",
    "        if isinstance(u, dict):\n",
    "            vals = []\n",
    "            for key in (\"Swiss-Prot\", \"TrEMBL\"):\n",
    "                v = u.get(key)\n",
    "                if isinstance(v, str):\n",
    "                    vals.append(v)\n",
    "                elif isinstance(v, list):\n",
    "                    vals.extend(v)\n",
    "            return vals\n",
    "        if isinstance(u, list):\n",
    "            # assume it's already a list of strings\n",
    "            return [x for x in u if isinstance(x, str)]\n",
    "        # anything else (NaN, etc.)\n",
    "        return []\n",
    "\n",
    "    df[\"uniprot_list\"] = df[\"uniprot\"].apply(normalize_uniprot)\n",
    "\n",
    "    # explode so each UniProt ID is on its own row\n",
    "    df = df.explode(\"uniprot_list\")\n",
    "\n",
    "    # drop empties\n",
    "    df = df.dropna(subset=[\"uniprot_list\"])\n",
    "\n",
    "    # deduplicate and return as a simple list\n",
    "    return sorted(df[\"uniprot_list\"].unique().tolist())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "294007a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "mapk_genes = symbols_to_uniprot(gene_sets[mapk_key])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5cf9d86b",
   "metadata": {},
   "outputs": [],
   "source": [
    "mapk_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "305e9006",
   "metadata": {},
   "outputs": [],
   "source": [
    "# mapk_uniprot is your UniProt list from symbols_to_uniprot(...)\n",
    "# make sure they exist in adata\n",
    "mapk_genes_in_data = sorted(set(mapk_genes).intersection(set(adata.var_names)))\n",
    "print(f\"{len(mapk_genes_in_data)} MAPK genes overlap with adata features.\")\n",
    "\n",
    "gene_sets = {\"MAPK\": mapk_genes_in_data}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "248abf95",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gseapy as gp\n",
    "\n",
    "# expression matrix: samples x features\n",
    "expr = adata.to_df()      # uses .X by default\n",
    "# transpose to features x samples for gseapy\n",
    "expr_t = expr.T\n",
    "\n",
    "ss = gp.ssgsea(\n",
    "    data=expr_t,\n",
    "    gene_sets=gene_sets,   # {\"MAPK\": [UniProt IDs]}\n",
    "    outdir=None,           # no output to disk\n",
    "    sample_norm=True,\n",
    "    min_size=1,            # or 5, but 1 is safer if overlap is small\n",
    "    max_size=5000,\n",
    "    permutation_num=0      # standard for ssGSEA\n",
    ")\n",
    "\n",
    "scores_df = ss.res2d       # rows = pathways, cols = samples\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "05229732",
   "metadata": {},
   "outputs": [],
   "source": [
    "scores_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9242e246",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1. Make a mapping: filename → NES\n",
    "nes_map = dict(zip(scores_df[\"Name\"], scores_df[\"NES\"]))\n",
    "\n",
    "# 2. Map it into adata.obs\n",
    "adata.obs[\"MAPK_ssGSEA\"] = adata.obs[\"filename\"].map(nes_map)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "455847d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b9818777",
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "df = adata.obs\n",
    "\n",
    "plt.figure(figsize=(8, 5))\n",
    "sns.boxplot(data=df, x=\"mapk\", y=\"MAPK_ssGSEA\")\n",
    "sns.stripplot(data=df, x=\"mapk\", y=\"MAPK_ssGSEA\", color=\"black\", alpha=0.5)\n",
    "\n",
    "plt.title(\"MAPK ssGSEA scores by mapk\")\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b95aae33",
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "\n",
    "df = adata.obs.copy()\n",
    "\n",
    "# ---- 1️⃣ Extract correct order & colors from adata.uns ----\n",
    "type_order = adata.uns[\"colors_type\"].keys()\n",
    "type_colors = adata.uns[\"colors_type\"].values()\n",
    "\n",
    "# Map type → color\n",
    "type_to_color = dict(zip(type_order, type_colors))\n",
    "\n",
    "# ---- 2️⃣ Sort samples by Type (and keep sample name index) ----\n",
    "df[\"type\"] = pd.Categorical(df[\"type\"], categories=type_order, ordered=True)\n",
    "df = df.sort_values(\"type\")\n",
    "\n",
    "# ---- 3️⃣ Create color array matching each sample ----\n",
    "sample_colors = df[\"type\"].map(type_to_color)\n",
    "\n",
    "# ---- 4️⃣ Bar plot (each bar = sample) ----\n",
    "plt.figure(figsize=(14, 5))\n",
    "plt.bar(\n",
    "    x=df.index,\n",
    "    height=df[\"MAPK_ssGSEA\"],\n",
    "    color=sample_colors\n",
    ")\n",
    "\n",
    "plt.xticks(rotation=90)\n",
    "plt.ylabel(\"MAPK ssGSEA\")\n",
    "plt.xlabel(\"Samples\")\n",
    "plt.title(\"MAPK ssGSEA per sample grouped by type\")\n",
    "plt.tight_layout()\n",
    "\n",
    "# Add legend manually\n",
    "handles = [plt.Rectangle((0,0),1,1, color=type_to_color[t]) for t in type_order]\n",
    "plt.legend(handles, type_order, title=\"Type\", bbox_to_anchor=(1.05, 1), loc=\"upper left\")\n",
    "\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "84a00493",
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.stats import ranksums\n",
    "\n",
    "group1 = adata.obs.loc[adata.obs[\"mapk\"] == \"positive\", \"MAPK_ssGSEA\"]\n",
    "group2 = adata.obs.loc[adata.obs[\"mapk\"] == \"negative\", \"MAPK_ssGSEA\"]\n",
    "\n",
    "stat, pval = ranksums(group1, group2)\n",
    "print(\"Wilcoxon rank-sum p-value:\", pval)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d63a0afa",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c514b337",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "47549a84",
   "metadata": {},
   "outputs": [],
   "source": [
    "res = ttest_ind_all_vars(\n",
    "    adata,\n",
    "    group1_query='mapk == \"positive\"',\n",
    "    group2_query='mapk == \"negative\"'\n",
    ")\n",
    "\n",
    "res[(res[\"p_adj\"] < 0.05) & (abs(res[\"log2FC\"]) > 0.4)] #.sort_values(\"log2FC\", ascending=False).head(10)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "309a7aa7",
   "metadata": {},
   "outputs": [],
   "source": [
    "volcano_plot(res)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "90b22668",
   "metadata": {},
   "outputs": [],
   "source": [
    "res.loc[\"A0A024RBG1\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d31436b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "sp.stats.ttest_ind(\n",
    "    adata[adata.obs[\"type\"] == \"MNG\", \"A0A024RBG1\"].X.toarray().flatten(),\n",
    "    adata[adata.obs[\"type\"] == \"GG\", \"A0A024RBG1\"].X.toarray().flatten(),\n",
    "    #equal_var=False,\n",
    "    nan_policy='omit'\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "774fb1c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7dd78b29",
   "metadata": {},
   "outputs": [],
   "source": [
    "res = group_difference_with_covariates(\n",
    "    adata,\n",
    "    group=\"sex\",\n",
    "    covariates=[\"type\",\"age\"],\n",
    ")\n",
    "\n",
    "res.sort_values(\"p_adj\").head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "70bc8798",
   "metadata": {},
   "outputs": [],
   "source": [
    "res[res[\"p_adj\"] < 0.05]\n",
    "# all 3 on y chromosome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c889e8e",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_sub = adata[(adata.obs[\"mapk\"] == \"pos\") | (adata.obs[\"mapk\"] == \"neg\")].copy()\n",
    "\n",
    "res = group_difference_with_covariates(\n",
    "    adata_sub,\n",
    "    group=\"mapk\",\n",
    "    covariates=[\"type\",\"age\",\"sex\"],\n",
    ")\n",
    "\n",
    "res.sort_values(\"p_adj\").head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61702771",
   "metadata": {},
   "outputs": [],
   "source": [
    "res[res[\"p_adj\"] < 0.01]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "07880d8b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e8360a97",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a58546e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ebafd3ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "from scipy import stats\n",
    "\n",
    "def anova_all_vars(\n",
    "    adata,\n",
    "    groupby: str,\n",
    "    groups=None,\n",
    "    strict_na: bool = True,\n",
    "):\n",
    "    \"\"\"\n",
    "    One-way ANOVA for all variables in `adata.X` across groups defined in `adata.obs[groupby]`.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    adata : AnnData\n",
    "        Input AnnData object.\n",
    "    groupby : str\n",
    "        Column name in adata.obs defining the groups (e.g. 'cell_type').\n",
    "    groups : list-like, optional\n",
    "        Specific group labels to include (subset/order). By default, use all unique values.\n",
    "    strict_na : bool, default True\n",
    "        If True, raise ValueError if any NA is present in the data.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    pandas.DataFrame\n",
    "        Index: adata.var_names\n",
    "        Columns:\n",
    "            - F        : F-statistic\n",
    "            - p_val    : raw p-value\n",
    "            - p_adj    : BH-FDR\n",
    "            - mean_<g> : mean expression in group g for each variable\n",
    "    \"\"\"\n",
    "\n",
    "    obs_col = adata.obs[groupby]\n",
    "\n",
    "    if groups is None:\n",
    "        # preserve order of appearance\n",
    "        groups = pd.unique(obs_col)\n",
    "    else:\n",
    "        groups = pd.Index(groups)\n",
    "\n",
    "    # collect group matrices, sizes, and means\n",
    "    X = adata.X\n",
    "    group_mats = []\n",
    "    group_sizes = []\n",
    "    group_means = []\n",
    "    group_ss_within = []\n",
    "\n",
    "    for g in groups:\n",
    "        idx = (obs_col == g).values\n",
    "        if not np.any(idx):\n",
    "            raise ValueError(f\"No observations found for group '{g}' in column '{groupby}'.\")\n",
    "\n",
    "        Xg = X[idx, :]\n",
    "        # to dense if sparse\n",
    "        if not isinstance(Xg, np.ndarray):\n",
    "            Xg = Xg.toarray()\n",
    "\n",
    "        if strict_na and np.isnan(Xg).any():\n",
    "            raise ValueError(\n",
    "                f\"Input expression matrix for group '{g}' contains NA values — clean your data first.\"\n",
    "            )\n",
    "\n",
    "        n_g = Xg.shape[0]\n",
    "        mean_g = Xg.mean(axis=0)\n",
    "\n",
    "        # within-group SS: sum (x_ig - mean_g)^2\n",
    "        ss_within_g = ((Xg - mean_g) ** 2).sum(axis=0)\n",
    "\n",
    "        group_mats.append(Xg)\n",
    "        group_sizes.append(n_g)\n",
    "        group_means.append(mean_g)\n",
    "        group_ss_within.append(ss_within_g)\n",
    "\n",
    "    group_sizes = np.array(group_sizes)[:, None]   # shape (k, 1)\n",
    "    group_means = np.vstack(group_means)          # shape (k, n_vars)\n",
    "    group_ss_within = np.vstack(group_ss_within)  # shape (k, n_vars)\n",
    "\n",
    "    # total N and grand mean per variable\n",
    "    N_total = group_sizes.sum(axis=0)[0]\n",
    "    grand_mean = (group_sizes * group_means).sum(axis=0) / N_total\n",
    "\n",
    "    # SS_between and SS_within\n",
    "    ss_between = (group_sizes * (group_means - grand_mean) ** 2).sum(axis=0)\n",
    "    ss_within = group_ss_within.sum(axis=0)\n",
    "\n",
    "    k = len(groups)\n",
    "    df_between = k - 1\n",
    "    df_within = N_total - k\n",
    "\n",
    "    # mean squares\n",
    "    ms_between = ss_between / df_between\n",
    "    ms_within = ss_within / df_within\n",
    "\n",
    "    # avoid division by zero: where ms_within == 0, set F and p to NaN\n",
    "    with np.errstate(divide=\"ignore\", invalid=\"ignore\"):\n",
    "        F = ms_between / ms_within\n",
    "    F[~np.isfinite(F)] = np.nan\n",
    "\n",
    "    # p-values from F-distribution\n",
    "    p_vals = stats.f.sf(F, df_between, df_within)\n",
    "    p_vals[np.isnan(F)] = np.nan\n",
    "\n",
    "    # Benjamini–Hochberg FDR (ignore NaNs in ordering)\n",
    "    p = p_vals.copy()\n",
    "    n = p.size\n",
    "\n",
    "    # treat NaNs as 1.0 for sorting so they go to the end\n",
    "    p_for_sort = np.nan_to_num(p, nan=1.0)\n",
    "    order = np.argsort(p_for_sort)\n",
    "    ranks = np.arange(1, n + 1, dtype=float)\n",
    "\n",
    "    bh = (p_for_sort[order] * n) / ranks\n",
    "    bh = np.minimum.accumulate(bh[::-1])[::-1]\n",
    "\n",
    "    p_adj = np.empty_like(bh)\n",
    "    p_adj[order] = bh\n",
    "\n",
    "    # build result DataFrame\n",
    "    data = {\n",
    "        \"F\": F,\n",
    "        \"p_val\": p_vals,\n",
    "        \"p_adj\": p_adj,\n",
    "    }\n",
    "\n",
    "    # add per-group means: mean_<group>\n",
    "    for i, g in enumerate(groups):\n",
    "        data[f\"mean_{g}\"] = group_means[i, :]\n",
    "\n",
    "    res = pd.DataFrame(data, index=adata.var_names)\n",
    "\n",
    "    return res\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da6f4e34",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_anova = anova_all_vars(\n",
    "    adata,\n",
    "    groupby=\"Type\",\n",
    "    # optionally: groups=[\"Progenitor\", \"Ortho\", \"OtherType\"]\n",
    ")\n",
    "\n",
    "# top ANOVA hits\n",
    "res_anova.sort_values(\"p_adj\").head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3706ff3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f6b5333d",
   "metadata": {},
   "outputs": [],
   "source": [
    "sp.stats.ttest_ind(\n",
    "    adata[adata.obs[\"cell_type\"] == \"Ortho\", \"A0A024RBG1\"].X.toarray().flatten(),\n",
    "    adata[adata.obs[\"cell_type\"] == \"Progenitor\", \"A0A024RBG1\"].X.toarray().flatten(),\n",
    "    equal_var=False,\n",
    "    nan_policy='omit'\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa87f9ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "ttest_groups_equalvar(adata, groupby=\"cell_type\", group1=\"Progenitor\", group2=\"Ortho\", layer=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4ac6c708",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "404b2a53",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.rank_genes_groups(\n",
    "        adata,\n",
    "        groupby=\"cell_type\",\n",
    "        groups=[\"Ortho\"],\n",
    "        reference=\"Progenitor\",   \n",
    "        method=\"t-test\",\n",
    "        corr_method=\"benjamini-hochberg\",\n",
    "        n_jobs=1\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01119419",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.uns['rank_genes_groups']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9310e3d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "diff_Ortho_Progenitor = sc.get.rank_genes_groups_df(adata, group=\"Ortho\", gene_symbols=\"gene_name\")\n",
    "diff_Ortho_Progenitor.sort_values(by=\"logfoldchanges\", ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c492dd2",
   "metadata": {},
   "outputs": [],
   "source": [
    "diff_Ortho_Progenitor[diff_Ortho_Progenitor[\"names\"].str.contains(\"P02730\", case=False, na=False)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "09fdcd53",
   "metadata": {},
   "outputs": [],
   "source": [
    "diff_Ortho_Progenitor[diff_Ortho_Progenitor[\"names\"].str.contains(\"Q08495\", case=False, na=False)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "58132f44",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "def volcano_plot(\n",
    "    res: pd.DataFrame,\n",
    "    fc_col: str = \"log2fc\",\n",
    "    p_col: str = \"qval\",          # will fall back to 'pval' if missing\n",
    "    fc_thresh: float = 1.0,       # |log2FC| cutoff\n",
    "    p_thresh: float = 0.05,       # FDR (or p) cutoff\n",
    "    label_top: int = 10,          # annotate top N most significant hits\n",
    "    title: str | None = None,\n",
    "    ax: bool = False,\n",
    "    show: bool = True,\n",
    "    save: str | None = None,\n",
    "):\n",
    "    \"\"\"\n",
    "    Simple volcano plot:\n",
    "      x = log2 fold-change\n",
    "      y = -log10(q-value) (or p-value if qval not present)\n",
    "\n",
    "    Expects columns:\n",
    "      - fc_col (default 'log2fc')\n",
    "      - p_col (default 'qval', falls back to 'pval' if not found)\n",
    "      - 'protein_id' for labels (falls back to 'names' or index)\n",
    "    \"\"\"\n",
    "    df = res.copy()\n",
    "\n",
    "    # choose p column\n",
    "    if p_col not in df.columns:\n",
    "        if \"pval\" in df.columns:\n",
    "            p_col = \"pval\"\n",
    "        else:\n",
    "            raise ValueError(f\"Neither '{p_col}' nor 'pval' found in columns: {list(df.columns)}\")\n",
    "\n",
    "    # coerce numeric and clean\n",
    "    df[fc_col] = pd.to_numeric(df[fc_col], errors=\"coerce\")\n",
    "    df[p_col]  = pd.to_numeric(df[p_col],  errors=\"coerce\")\n",
    "\n",
    "    # guard against zeros -> inf\n",
    "    eps = 1e-300\n",
    "    p_vals = df[p_col].clip(lower=eps)\n",
    "    df[\"_mlog10p\"] = -np.log10(p_vals)\n",
    "\n",
    "    # masks\n",
    "    sig_mask  = (p_vals <= p_thresh) & (df[fc_col].abs() >= fc_thresh)\n",
    "    up_mask   = sig_mask & (df[fc_col] > 0)\n",
    "    down_mask = sig_mask & (df[fc_col] < 0)\n",
    "    ns_mask   = ~sig_mask\n",
    "\n",
    "    # axis\n",
    "    fig, ax = plt.subplots(figsize=(6, 5))\n",
    "\n",
    "    # scatter (no explicit colors -> matplotlib defaults)\n",
    "    ax.scatter(df.loc[ns_mask, fc_col],   df.loc[ns_mask, \"_mlog10p\"], s=10, alpha=0.3,c='#B2AEA9')\n",
    "    ax.scatter(df.loc[up_mask, fc_col],   df.loc[up_mask, \"_mlog10p\"], s=14, alpha=0.4, c='#FF4C00')\n",
    "    ax.scatter(df.loc[down_mask, fc_col], df.loc[down_mask, \"_mlog10p\"], s=14, alpha=0.4, c='#466CE7')\n",
    "\n",
    "    # thresholds\n",
    "    ax.axvline(+fc_thresh, linestyle=\"--\", linewidth=1, c='gray')\n",
    "    ax.axvline(-fc_thresh, linestyle=\"--\", linewidth=1, c='gray')\n",
    "    ax.axhline(-np.log10(p_thresh), linestyle=\"--\", linewidth=1, c='gray')\n",
    "\n",
    "    # labels & title\n",
    "    ax.set_xlabel(\"log2 fold change\")\n",
    "    ax.set_ylabel(\"-log10(p)\")\n",
    "    if title:\n",
    "        ax.set_title(title)\n",
    "\n",
    "    # annotate top hits by significance\n",
    "    name_col = \"gene_name\" if \"gene_name\" in df.columns else (\"names\" if \"names\" in df.columns else None)\n",
    "    if label_top and name_col:\n",
    "        top = df.loc[sig_mask].sort_values(by=[p_col, fc_col], ascending=[True, False]).head(label_top)\n",
    "        for _, r in top.iterrows():\n",
    "            ax.annotate(\n",
    "                str(r[name_col]),\n",
    "                (r[fc_col], r[\"_mlog10p\"]),\n",
    "                xytext=(3, 3),\n",
    "                textcoords=\"offset points\",\n",
    "                fontsize=8\n",
    "            )\n",
    "\n",
    "    # make x-lims symmetric around 0 for a nicer look\n",
    "    xmax = np.nanpercentile(df[fc_col].abs(), 99)\n",
    "    if np.isfinite(xmax) and xmax > 0:\n",
    "        ax.set_xlim(-xmax, xmax)\n",
    "\n",
    "    plt.tight_layout()\n",
    "    if save:\n",
    "        plt.savefig(save, dpi=300, bbox_inches=\"tight\")\n",
    "    if show:\n",
    "        plt.show()\n",
    "    if ax:\n",
    "        return ax\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "291b729d",
   "metadata": {},
   "outputs": [],
   "source": [
    "volcano_plot(\n",
    "    res = diff_Ortho_Progenitor,\n",
    "    fc_col = \"logfoldchanges\",\n",
    "    p_col = \"pvals_adj\",         \n",
    "    fc_thresh = 1.0,  \n",
    "    p_thresh = 0.01,\n",
    "    label_top=50,\n",
    "    title=\"Ortho vs Progenitor Volcano Plot\"\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c10726c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "prot_diff_Ortho_Progenitor_up = diff_Ortho_Progenitor[(diff_Ortho_Progenitor['pvals_adj'] < 0.01) & (diff_Ortho_Progenitor['logfoldchanges'] > 1)].names.tolist()\n",
    "prot_diff_Ortho_Progenitor_up = [x.split(\";\")[0].strip() for x in prot_diff_Ortho_Progenitor_up]\n",
    "pd.Series(prot_diff_Ortho_Progenitor_up).to_csv(\"/Users/isabell/Documents/projects/protypy/Karayel2020/prot_diff_Ortho_Progenitor_up.csv\", index=False)\n",
    "len(prot_diff_Ortho_Progenitor_up)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18064ff3",
   "metadata": {},
   "outputs": [],
   "source": [
    "prot_diff_Ortho_Progenitor_down = diff_Ortho_Progenitor[(diff_Ortho_Progenitor['pvals_adj'] < 0.01) & (diff_Ortho_Progenitor['logfoldchanges'] < -1)].names.tolist()\n",
    "prot_diff_Ortho_Progenitor_down = [x.split(\";\")[0].strip() for x in prot_diff_Ortho_Progenitor_down]\n",
    "pd.Series(prot_diff_Ortho_Progenitor_down).to_csv(\"/Users/isabell/Documents/projects/protypy/Karayel2020/prot_diff_Ortho_Progenitor_down.csv\", index=False)\n",
    "len(prot_diff_Ortho_Progenitor_down)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "84c80e8f",
   "metadata": {},
   "source": [
    "## ANOVA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ca254f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import statsmodels.formula.api as smf\n",
    "from statsmodels.stats.anova import anova_lm\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "from statsmodels.stats.multicomp import pairwise_tukeyhsd\n",
    "import scanpy as sc\n",
    "\n",
    "def anova_statsmodels(adata, groupby, covariates=None, layer=None, typ=2, q_thresh=0.05):\n",
    "    \"\"\"\n",
    "    One-way (or ANCOVA) per-feature ANOVA using statsmodels.\n",
    "    Returns an ANOVA table plus post-hoc Tukey HSD for significant features.\n",
    "    \"\"\"\n",
    "    covariates = covariates or []\n",
    "\n",
    "    # matrix\n",
    "    X = adata.layers[layer] if layer is not None else adata.X\n",
    "    X = X.A if hasattr(X, \"A\") else (X.toarray() if hasattr(X, \"toarray\") else X)\n",
    "    df_obs = adata.obs[covariates + [groupby]].copy()\n",
    "    df_obs[groupby] = df_obs[groupby].astype(\"category\")\n",
    "\n",
    "    # ---- ANOVA per feature\n",
    "    rows = []\n",
    "    for j, feat in enumerate(adata.var_names):\n",
    "        df = df_obs.copy()\n",
    "        df[\"y\"] = X[:, j]\n",
    "        rhs = \"C(\" + groupby + \")\" + (\" + \" + \" + \".join(covariates) if covariates else \"\")\n",
    "        try:\n",
    "            model = smf.ols(f\"y ~ {rhs}\", data=df, missing=\"drop\").fit()\n",
    "            aov = anova_lm(model, typ=typ)\n",
    "            fval = aov.loc[f\"C({groupby})\", \"F\"]\n",
    "            pval = aov.loc[f\"C({groupby})\", \"PR(>F)\"]\n",
    "        except Exception:\n",
    "            fval, pval = np.nan, np.nan\n",
    "        rows.append((feat, fval, pval))\n",
    "\n",
    "    aov_res = pd.DataFrame(rows, columns=[\"feature\", \"F\", \"pval\"])\n",
    "    aov_res[\"qval\"] = multipletests(aov_res[\"pval\"], method=\"fdr_bh\")[1]\n",
    "    aov_res = aov_res.sort_values(\"qval\").reset_index(drop=True)\n",
    "\n",
    "    # ---- Post-hoc: Tukey HSD on significant features\n",
    "    sig_feats = aov_res.loc[aov_res[\"qval\"] <= q_thresh, \"feature\"].tolist()\n",
    "    tukey_rows = []\n",
    "    if len(sig_feats):\n",
    "        groups = df_obs[groupby].astype(\"category\")\n",
    "        cats = list(groups.cat.categories)\n",
    "\n",
    "        for feat in sig_feats:\n",
    "            y = X[:, adata.var_names.get_loc(feat)]\n",
    "            g = groups  # same for all features\n",
    "            # Tukey works on arrays (ignores covariates, i.e. pure one-way groups)\n",
    "            try:\n",
    "                tk = pairwise_tukeyhsd(endog=y, groups=g, alpha=0.05)\n",
    "                tk_df = pd.DataFrame(\n",
    "                    tk._results_table.data[1:], columns=tk._results_table.data[0]\n",
    "                )\n",
    "                # normalize column names\n",
    "                tk_df.columns = [\"group1\", \"group2\", \"meandiff\", \"p_adj\", \"lower\", \"upper\", \"reject\"]\n",
    "                tk_df.insert(0, \"feature\", feat)\n",
    "                # Add direction (which group has higher mean?)\n",
    "                # Compute group means for feature:\n",
    "                means = pd.Series({cat: np.nanmean(y[g == cat]) for cat in cats})\n",
    "                tk_df[\"mean_g1\"] = tk_df[\"group1\"].map(means)\n",
    "                tk_df[\"mean_g2\"] = tk_df[\"group2\"].map(means)\n",
    "                tk_df[\"direction\"] = np.where(tk_df[\"meandiff\"] > 0, \"group1>group2\", \"group2>group1\")\n",
    "                tukey_rows.append(tk_df)\n",
    "            except Exception:\n",
    "                pass\n",
    "\n",
    "    tukey_res = pd.concat(tukey_rows, ignore_index=True) if tukey_rows else pd.DataFrame(\n",
    "        columns=[\"feature\",\"group1\",\"group2\",\"meandiff\",\"p_adj\",\"lower\",\"upper\",\"reject\",\"mean_g1\",\"mean_g2\",\"direction\"]\n",
    "    )\n",
    "\n",
    "    return aov_res, tukey_res\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e598269",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1) Run ANOVA + post-hoc\n",
    "aov_res, tukey_res = anova_statsmodels(\n",
    "    adata,\n",
    "    groupby=\"cell_type\",     # your grouping in adata.obs\n",
    "    covariates=None,    # optional; [] if none\n",
    "    layer=None,              # or a layer name like \"log1p\"\n",
    "    typ=2,                   # Type II SS (robust with imbalance)\n",
    "    q_thresh=0.05\n",
    ")\n",
    "\n",
    "# 2) Top ANOVA hits\n",
    "aov_res.head()\n",
    "\n",
    "# 3) Post-hoc table: which groups differ, by how much, and in which direction\n",
    "tukey_res.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b94a306b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "08fbb138",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a040d3a3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d6f609a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "from scipy import sparse\n",
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "\n",
    "def diff_proteins_scanpy(\n",
    "    adata,\n",
    "    groupby: str,           # e.g. 'tumor_type' in adata.obs\n",
    "    group1: str,            # e.g. 'GBM'\n",
    "    group2: str = 'rest',   # or a specific label, e.g. 'LGG'\n",
    "    method: str = 'wilcoxon',   # 't-test', 'wilcoxon', 'logreg', 't-test_overestim_var'\n",
    "    layer: str | None = None,   # choose a layer or use .X\n",
    "    impute: str = 'col_median'  # or 'none'\n",
    ") -> pd.DataFrame:\n",
    "    \"\"\"\n",
    "    Differential analysis on AnnData using Scanpy.\n",
    "    Returns a DataFrame with protein_id, log2fc, stat, pval, qval, and n_in/n_out.\n",
    "    \"\"\"\n",
    "    ad = adata.copy()\n",
    "    if layer is not None:\n",
    "        ad.X = adata.layers[layer]\n",
    "\n",
    "    # ensure dense float and handle NaNs simply (Scanpy tests expect finite values)\n",
    "    X = ad.X.toarray() if sparse.issparse(ad.X) else np.asarray(ad.X, dtype=float)\n",
    "    if impute == 'col_median':\n",
    "        col_med = np.nanmedian(X, axis=0)\n",
    "        r, c = np.where(~np.isfinite(X))\n",
    "        if r.size:\n",
    "            X[r, c] = col_med[c]\n",
    "    elif impute == 'none':\n",
    "        # replace remaining non-finite with 0 just to avoid crashes (conservative)\n",
    "        X[~np.isfinite(X)] = 0.0\n",
    "    ad.X = X\n",
    "\n",
    "    # run DE\n",
    "    sc.tl.rank_genes_groups(\n",
    "        ad,\n",
    "        groupby=groupby,\n",
    "        groups=[group1],\n",
    "        reference=group2,      # 'rest' or a specific label\n",
    "        method=method,\n",
    "        n_jobs=1\n",
    "    )\n",
    "\n",
    "    # collect results\n",
    "    df = sc.get.rank_genes_groups_df(ad, group=group1)\n",
    "    df = df.rename(columns={\n",
    "        'names': 'protein_id',\n",
    "        'logfoldchanges': 'log2fc',\n",
    "        'scores': 'stat',\n",
    "        'pvals': 'pval',\n",
    "        'pvals_adj': 'qval'\n",
    "    })\n",
    "    # add sample counts for each side (helpful QC)\n",
    "    n_in = (ad.obs[groupby] == group1).sum()\n",
    "    n_out = (ad.obs[groupby] != group1).sum() if group2 == 'rest' else (ad.obs[groupby] == group2).sum()\n",
    "    df['n_in'] = n_in\n",
    "    df['n_out'] = n_out\n",
    "    return df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "afac7ad9",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_n_p1 = diff_proteins_scanpy(\n",
    "    adata, \n",
    "    groupby='cell_type', \n",
    "    group1='Negativefrac', \n",
    "    group2='P1andP2', \n",
    "    method='t-test')\n",
    "\n",
    "res_n_p1.sort_values(by='qval', ascending=True)[0:20]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e369a3a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "volcano_plot(res_n_p1, fc_thresh=1.0, p_thresh=0.05, label_top=5, title=\"Negativefrac vs P1andP2 (t-test)\") "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "577ba7a3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17af59fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "from scipy import sparse\n",
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "\n",
    "def diff_proteins_scanpy(\n",
    "    adata,\n",
    "    groupby: str,           # e.g. 'tumor_type' in adata.obs\n",
    "    group1: str,            # e.g. 'GBM'\n",
    "    group2: str = 'rest',   # or a specific label, e.g. 'LGG'\n",
    "    method: str = 'wilcoxon',   # 't-test', 'wilcoxon', 'logreg', 't-test_overestim_var'\n",
    "    layer: str | None = None,   # choose a layer or use .X\n",
    "    impute: str = 'col_median'  # or 'none'\n",
    ") -> pd.DataFrame:\n",
    "    \"\"\"\n",
    "    Differential analysis on AnnData using Scanpy.\n",
    "    Returns a DataFrame with protein_id, log2fc, stat, pval, qval, and n_in/n_out.\n",
    "    \"\"\"\n",
    "    ad = adata.copy()\n",
    "    if layer is not None:\n",
    "        ad.X = adata.layers[layer]\n",
    "\n",
    "    # ensure dense float and handle NaNs simply (Scanpy tests expect finite values)\n",
    "    X = ad.X.toarray() if sparse.issparse(ad.X) else np.asarray(ad.X, dtype=float)\n",
    "    if impute == 'col_median':\n",
    "        col_med = np.nanmedian(X, axis=0)\n",
    "        r, c = np.where(~np.isfinite(X))\n",
    "        if r.size:\n",
    "            X[r, c] = col_med[c]\n",
    "    elif impute == 'none':\n",
    "        # replace remaining non-finite with 0 just to avoid crashes (conservative)\n",
    "        X[~np.isfinite(X)] = 0.0\n",
    "    ad.X = X\n",
    "\n",
    "    # run DE\n",
    "    sc.tl.rank_genes_groups(\n",
    "        ad,\n",
    "        groupby=groupby,\n",
    "        groups=[group1],\n",
    "        reference=group2,      # 'rest' or a specific label\n",
    "        method=method,\n",
    "        n_jobs=1\n",
    "    )\n",
    "\n",
    "    # collect results\n",
    "    df = sc.get.rank_genes_groups_df(ad, group=group1)\n",
    "    df = df.rename(columns={\n",
    "        'names': 'protein_id',\n",
    "        'logfoldchanges': 'log2fc',\n",
    "        'scores': 'stat',\n",
    "        'pvals': 'pval',\n",
    "        'pvals_adj': 'qval'\n",
    "    })\n",
    "    # add sample counts for each side (helpful QC)\n",
    "    n_in = (ad.obs[groupby] == group1).sum()\n",
    "    n_out = (ad.obs[groupby] != group1).sum() if group2 == 'rest' else (ad.obs[groupby] == group2).sum()\n",
    "    df['n_in'] = n_in\n",
    "    df['n_out'] = n_out\n",
    "    return df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "86b9b193",
   "metadata": {},
   "outputs": [],
   "source": [
    "diff_neg_p1 = diff_proteins_scanpy(\n",
    "    adata, \n",
    "    groupby='cell_type', \n",
    "    group1='Negativefrac', \n",
    "    group2='P1andP2', \n",
    "    method='t-test')\n",
    "\n",
    "diff_neg_p1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7125a3b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "volcano_plot(diff_neg_p1, \n",
    "             fc_col = \"log2fc\",\n",
    "             p_col = \"qval\",\n",
    "             fc_thresh=1.0, p_thresh=0.05, label_top=5, title=\"Negativefrac vs P1andP2 (t-test)\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c72bb6f5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "613008a7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "83593803",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e276c216",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.rank_genes_groups(adata, n_genes=25, sharey=False, key = \"t-test\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a90ff529",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.rank_genes_groups_heatmap(adata, n_genes=10, key=\"t-test\", groupby=\"cell_type\", show_gene_labels=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "213bb4da",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.rank_genes_groups_matrixplot(adata, n_genes=5, key=\"t-test\", groupby=\"cell_type\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a7aa33b",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.rank_genes_groups_dotplot(adata, n_genes=5, key=\"t-test\", groupby=\"cell_type\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ab8ba4e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a86ac93",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Install if needed\n",
    "!pip install h5py\n",
    "\n",
    "# Import in your Python script or notebook\n",
    "import h5py\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6764f16b",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = ad.read_h5ad(\"/Users/isabell/Downloads/adata_analyzed.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46c0d7eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c1cfc47",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e7268b50",
   "metadata": {},
   "outputs": [],
   "source": [
    "set(adata.var[adata.var.proteoform_score > 0.4].protein_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9721287c",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.var[adata.var['protein_id'] == 'Q9Y2J2'][[\"protein_id\",\"Genes\",\"Stripped.Sequence\",\"cluster_id\",\"proteoform_score\",\"proteoform_score_pval_adj\"]].sort_values(by=\"cluster_id\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4db572a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "prots_bludauI = ['Q9Y2J2']  # RNT4, LIMA1\n",
    "cp.pl.proteoform_intensities(\n",
    "    adata,\n",
    "    prots_bludauI,\n",
    "    group_by='tumor_family',\n",
    "    xlab_rotation=45,\n",
    "    #log_transform=True,\n",
    "    z_transform=False,\n",
    "    show_zeros=True,\n",
    "    group_by_label_rotation=30,\n",
    "    figsize=(100,10),\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4771190d",
   "metadata": {},
   "outputs": [],
   "source": [
    "prots_bludauI = ['Q9Y2J2']  # RNT4, LIMA1\n",
    "cp.pl.proteoform_intensities(\n",
    "    adata,\n",
    "    prots_bludauI,\n",
    "    group_by='batch_id', #'tumor_family',\n",
    "    xlab_rotation=45,\n",
    "    #log_transform=True,\n",
    "    z_transform=False,\n",
    "    show_zeros=True,\n",
    "    group_by_label_rotation=30,\n",
    "    figsize=(100,10),\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87bcdc26",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (copro-dev)",
   "language": "python",
   "name": "copro-dev"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
