{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "4ff7cecae9ce85aa",
   "metadata": {},
   "source": [
    "# Differential gene expression analysis\n",
    "\n",
    "To perform differential gene expression analysis we have several alternatives and modes:\n",
    "- Single-cell level\n",
    "- Pseudobulk level\n",
    "\n",
    "The `dotools_py` package includes two functions to automatically test for DEA between two conditions\n",
    "for all the celltypes we have defined in our object, as well a consensus function to run both approaches.\n",
    "\n",
    "## Environment setup"
   ]
  },
  {
   "cell_type": "code",
   "id": "6c786710cf60cc98",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-10-22T14:33:19.385589Z",
     "start_time": "2025-10-22T14:33:11.569958Z"
    }
   },
   "source": [
    "import anndata as ad\n",
    "import dotools_py as do\n",
    "import session_info\n",
    "\n",
    "\n",
    "adata = ad.read_h5ad(\"/Users/david/Downloads/Data10x/adata.h5ad\")\n",
    "adata"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-10-22 16:33:19,274 - Jupyter enviroment detected. Using \"inline\" backend\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 2783 × 18517\n",
       "    obs: 'batch', 'condition', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'total_counts_mt', 'log1p_total_counts_mt', 'pct_counts_mt', 'total_counts_ribo', 'log1p_total_counts_ribo', 'pct_counts_ribo', 'n_genes', 'n_counts', 'doublet_class', 'doublet_score', 'leiden', 'autoAnnot', 'celltypist_conf_score', 'annotation', 'annotation_recluster'\n",
       "    var: 'highly_variable', 'means', 'dispersions', 'dispersions_norm', 'highly_variable_nbatches', 'highly_variable_intersection'\n",
       "    uns: 'annotation_recluster_colors', 'hvg', 'leiden', 'log1p', 'neighbors', 'umap'\n",
       "    obsm: 'X_CCA', 'X_pca', 'X_umap'\n",
       "    layers: 'counts', 'logcounts'\n",
       "    obsp: 'connectivities', 'distances'"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 1
  },
  {
   "cell_type": "markdown",
   "id": "f7d2494b3a32b915",
   "metadata": {},
   "source": [
    "## DEA at the single-cell level\n",
    "\n",
    "Among the test we can use we have: wilcoxon, t-test, logistic regression, t-test with overestimation of the variance and the MAST test.\n",
    "The MAST test can be run directly using the `do.tl.run_mast`, the other tests can be run with `do.tl.rank_genes_groups`. Alternatively, we can use `do.tl.rank_genes_condition`, to use any of the test and automatically test for all the cell-types\n",
    "\n",
    "To reduce the computation time, we are going to only use the NK cells"
   ]
  },
  {
   "cell_type": "code",
   "id": "86d52fbd1e250d8e",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-10-22T14:35:28.466157Z",
     "start_time": "2025-10-22T14:34:01.780942Z"
    }
   },
   "source": [
    "nk = adata[adata.obs.annotation == \"NK\"].copy()\n",
    "\n",
    "df = do.tl.rank_genes_condition(\n",
    "    nk,\n",
    "    groupby=\"condition\",\n",
    "    subset_by=\"annotation\",\n",
    "    reference=\"healthy\",\n",
    "    groups=[\"disease\"],\n",
    "    method=\"mast\",\n",
    "    get_results=True,\n",
    ")"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-10-22 16:34:01,821 - Running DGEs for NK.\n",
      "2025-10-22 16:34:01,824 - Running MAST test\n",
      "2025-10-22 16:34:12,968 - Reference level is set to healthy\n"
     ]
    }
   ],
   "execution_count": 2
  },
  {
   "cell_type": "code",
   "id": "7c1655671242f65b",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-10-22T14:35:32.245768Z",
     "start_time": "2025-10-22T14:35:32.237350Z"
    }
   },
   "source": [
    "df.head(10)"
   ],
   "outputs": [
    {
     "data": {
      "text/plain": [
       "   GeneName     log2fc     pvals      padj  pts_group   pts_ref    group  \\\n",
       "0      A1BG   2.604576  0.001125  0.021286   0.200000  0.042553  disease   \n",
       "1  A1BG-AS1 -22.228531  0.781334  1.000000   0.000000  0.002660  disease   \n",
       "2       A2M -24.127653  0.639331  1.000000   0.000000  0.005319  disease   \n",
       "3   A2M-AS1  -0.216740  0.636112  1.000000   0.133333  0.170213  disease   \n",
       "4    A4GALT   0.000000  1.000000  1.000000   0.000000  0.000000  disease   \n",
       "5      AAAS  -1.100035  0.694980  1.000000   0.022222  0.047872  disease   \n",
       "6      AACS -25.872604  0.328578  0.901818   0.000000  0.015957  disease   \n",
       "7     AAED1   0.645098  0.001607  0.028256   0.044444  0.090426  disease   \n",
       "8     AAGAB   0.570027  0.643581  1.000000   0.088889  0.066489  disease   \n",
       "9      AAK1  -0.717432  0.129302  0.568715   0.333333  0.465426  disease   \n",
       "\n",
       "  annotation  \n",
       "0         NK  \n",
       "1         NK  \n",
       "2         NK  \n",
       "3         NK  \n",
       "4         NK  \n",
       "5         NK  \n",
       "6         NK  \n",
       "7         NK  \n",
       "8         NK  \n",
       "9         NK  "
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneName</th>\n",
       "      <th>log2fc</th>\n",
       "      <th>pvals</th>\n",
       "      <th>padj</th>\n",
       "      <th>pts_group</th>\n",
       "      <th>pts_ref</th>\n",
       "      <th>group</th>\n",
       "      <th>annotation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>2.604576</td>\n",
       "      <td>0.001125</td>\n",
       "      <td>0.021286</td>\n",
       "      <td>0.200000</td>\n",
       "      <td>0.042553</td>\n",
       "      <td>disease</td>\n",
       "      <td>NK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>-22.228531</td>\n",
       "      <td>0.781334</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.002660</td>\n",
       "      <td>disease</td>\n",
       "      <td>NK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A2M</td>\n",
       "      <td>-24.127653</td>\n",
       "      <td>0.639331</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.005319</td>\n",
       "      <td>disease</td>\n",
       "      <td>NK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A2M-AS1</td>\n",
       "      <td>-0.216740</td>\n",
       "      <td>0.636112</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.133333</td>\n",
       "      <td>0.170213</td>\n",
       "      <td>disease</td>\n",
       "      <td>NK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A4GALT</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>disease</td>\n",
       "      <td>NK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>AAAS</td>\n",
       "      <td>-1.100035</td>\n",
       "      <td>0.694980</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.022222</td>\n",
       "      <td>0.047872</td>\n",
       "      <td>disease</td>\n",
       "      <td>NK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>AACS</td>\n",
       "      <td>-25.872604</td>\n",
       "      <td>0.328578</td>\n",
       "      <td>0.901818</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.015957</td>\n",
       "      <td>disease</td>\n",
       "      <td>NK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>AAED1</td>\n",
       "      <td>0.645098</td>\n",
       "      <td>0.001607</td>\n",
       "      <td>0.028256</td>\n",
       "      <td>0.044444</td>\n",
       "      <td>0.090426</td>\n",
       "      <td>disease</td>\n",
       "      <td>NK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>AAGAB</td>\n",
       "      <td>0.570027</td>\n",
       "      <td>0.643581</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.088889</td>\n",
       "      <td>0.066489</td>\n",
       "      <td>disease</td>\n",
       "      <td>NK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>AAK1</td>\n",
       "      <td>-0.717432</td>\n",
       "      <td>0.129302</td>\n",
       "      <td>0.568715</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.465426</td>\n",
       "      <td>disease</td>\n",
       "      <td>NK</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 3
  },
  {
   "cell_type": "markdown",
   "id": "39d5e7b39f14dbd7",
   "metadata": {},
   "source": [
    "## DEA at the pseudobulk level\n",
    "\n",
    "To perform differential gene expression using a pseudobulk approach we can use `do.tl.rank_genes_pseudobulk`, which test between two conditions for each cell-type. We can use `DESEq2` or `edgeR`. In this case we need to generate pseudo-replicates since we only have one sample per condition."
   ]
  },
  {
   "cell_type": "code",
   "id": "2c670d6167bfd374",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-10-22T14:36:31.523200Z",
     "start_time": "2025-10-22T14:36:05.136712Z"
    }
   },
   "source": [
    "df = do.tl.rank_genes_pseudobulk(\n",
    "    adata,\n",
    "    ctrl_cond=\"healthy\",\n",
    "    disease_cond=\"disease\",\n",
    "    cluster_key=\"annotation\",\n",
    "    batch_key=\"batch\",\n",
    "    condition_key=\"condition\",\n",
    "    design=\"~condition\",\n",
    "    min_cells=30,\n",
    "    min_counts=10,\n",
    "    method=\"deseq2\",\n",
    "    pseudobulk_approach=\"sum\",\n",
    "    technical_replicates=2,\n",
    "    get_results=True,\n",
    ")"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-10-22 16:36:05,602 - Generating Pseudo-bulk data\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Pseudo-bulked groups: 100%|██████████| 10/10 [00:00<00:00, 54.47it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-10-22 16:36:23,693 - Removed 5806 genes for having less than 10 total counts\n",
      "2025-10-22 16:36:23,696 - Run DESeq2\n",
      "Using None as control genes, passed at DeseqDataSet initialization\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fitting size factors...\n",
      "... done in 0.00 seconds.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 0.34 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 0.07 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 0.48 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 0.40 seconds.\n",
      "\n",
      "Calculating cook's distance...\n",
      "... done in 0.00 seconds.\n",
      "\n",
      "Replacing 0 outlier genes.\n",
      "\n",
      "Running Wald tests...\n",
      "... done in 0.20 seconds.\n",
      "\n",
      "Fitting size factors...\n",
      "... done in 0.00 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: condition disease vs healthy\n",
      "          baseMean  log2FoldChange     lfcSE      stat    pvalue  padj\n",
      "A1BG      5.669955        0.403825  1.067884  0.378154  0.705316   NaN\n",
      "A1BG-AS1  0.738919        2.278757  3.406331  0.668977  0.503510   NaN\n",
      "A2M-AS1   0.157668        0.021039  5.332329  0.003946  0.996852   NaN\n",
      "A4GALT    2.649744        4.131701  2.716025  1.521231  0.128202   NaN\n",
      "AAAS      1.998178       -0.768574  1.708237 -0.449923  0.652766   NaN\n",
      "...            ...             ...       ...       ...       ...   ...\n",
      "ZXDB      1.219480       -4.522738  3.132528 -1.443798  0.148796   NaN\n",
      "ZXDC      3.052221       -0.368758  1.482192 -0.248793  0.803521   NaN\n",
      "ZYG11B    2.702957        1.185403  1.593806  0.743756  0.457024   NaN\n",
      "ZYX       4.646288        0.371801  1.177736  0.315692  0.752237   NaN\n",
      "ZZEF1     2.725794       -0.534042  1.487064 -0.359125  0.719501   NaN\n",
      "\n",
      "[12711 rows x 6 columns]\n",
      "Using None as control genes, passed at DeseqDataSet initialization\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fitting dispersions...\n",
      "... done in 0.35 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 0.07 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 0.46 seconds.\n",
      "\n",
      "Fitting LFCs...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-10-22 16:36:27,369 - Test could not be computed for Monocytes due to You specified a non-existant category for condition. Possible categories: healthy\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... done in 0.42 seconds.\n",
      "\n",
      "Calculating cook's distance...\n",
      "... done in 0.00 seconds.\n",
      "\n",
      "Replacing 0 outlier genes.\n",
      "\n",
      "Fitting size factors...\n",
      "... done in 0.00 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using None as control genes, passed at DeseqDataSet initialization\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fitting dispersions...\n",
      "... done in 0.42 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 0.07 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 0.61 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 0.49 seconds.\n",
      "\n",
      "Calculating cook's distance...\n",
      "... done in 0.00 seconds.\n",
      "\n",
      "Replacing 0 outlier genes.\n",
      "\n",
      "Running Wald tests...\n",
      "... done in 0.19 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: condition disease vs healthy\n",
      "           baseMean  log2FoldChange     lfcSE      stat    pvalue      padj\n",
      "A1BG       9.584556        2.614643  1.247043  2.096674  0.036022  0.102957\n",
      "A1BG-AS1   0.083831        1.607201  5.343868  0.300756  0.763600       NaN\n",
      "A2M-AS1   10.868651       -0.295972  0.595112 -0.497339  0.618950  0.740909\n",
      "A4GALT     0.000000             NaN       NaN       NaN       NaN       NaN\n",
      "AAAS       2.234670       -1.123438  1.523501 -0.737405  0.460876       NaN\n",
      "...             ...             ...       ...       ...       ...       ...\n",
      "ZXDB       0.337953       -0.396078  2.704777 -0.146436  0.883577       NaN\n",
      "ZXDC       1.524730       -2.568738  2.527960 -1.016131  0.309567       NaN\n",
      "ZYG11B     0.252808        0.021401  2.999455  0.007135  0.994307       NaN\n",
      "ZYX       11.933621       -1.723495  0.810586 -2.126234  0.033484  0.098208\n",
      "ZZEF1      3.545880       -3.787932  2.403851 -1.575777  0.115077       NaN\n",
      "\n",
      "[12711 rows x 6 columns]\n",
      "Using None as control genes, passed at DeseqDataSet initialization\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fitting size factors...\n",
      "... done in 0.00 seconds.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 0.40 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 0.08 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 0.51 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 0.35 seconds.\n",
      "\n",
      "Calculating cook's distance...\n",
      "... done in 0.00 seconds.\n",
      "\n",
      "Replacing 0 outlier genes.\n",
      "\n",
      "Running Wald tests...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: condition disease vs healthy\n",
      "           baseMean  log2FoldChange     lfcSE      stat    pvalue      padj\n",
      "A1BG      41.778608       -0.274318  0.380225 -0.721461  0.470626  0.670616\n",
      "A1BG-AS1   4.012429       -0.926947  1.281293 -0.723446  0.469406  0.669769\n",
      "A2M-AS1    4.688797       -1.240566  1.236603 -1.003204  0.315762  0.530839\n",
      "A4GALT     0.000000             NaN       NaN       NaN       NaN       NaN\n",
      "AAAS      15.380437       -1.665547  0.709661 -2.346961  0.018927  0.073339\n",
      "...             ...             ...       ...       ...       ...       ...\n",
      "ZXDB       5.162987        1.357682  1.049996  1.293036  0.195999  0.389684\n",
      "ZXDC      18.622773       -0.592434  0.604882 -0.979422  0.327372  0.542457\n",
      "ZYG11B    18.725857       -0.319720  0.563125 -0.567761  0.570197  0.748763\n",
      "ZYX       59.401751       -0.652206  0.324146 -2.012075  0.044212  0.138577\n",
      "ZZEF1     23.713091       -0.187387  0.493061 -0.380049  0.703909  0.841674\n",
      "\n",
      "[12711 rows x 6 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... done in 0.31 seconds.\n",
      "\n"
     ]
    }
   ],
   "execution_count": 4
  },
  {
   "cell_type": "code",
   "id": "c403288efb72ff1e",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-10-22T14:36:44.530997Z",
     "start_time": "2025-10-22T14:36:44.524520Z"
    }
   },
   "source": [
    "df.head(10)"
   ],
   "outputs": [
    {
     "data": {
      "text/plain": [
       "          baseMean  log2FoldChange     lfcSE      stat    pvalue      padj  \\\n",
       "A1BG      5.669955        0.403825  1.067884  0.378154  0.705316  1.000000   \n",
       "A1BG-AS1  0.738919        2.278757  3.406331  0.668977  0.503510  1.000000   \n",
       "A2M-AS1   0.157668        0.021039  5.332329  0.003946  0.996852  1.000000   \n",
       "A4GALT    2.649744        4.131701  2.716025  1.521231  0.128202  1.000000   \n",
       "AAAS      1.998178       -0.768574  1.708237 -0.449923  0.652766  1.000000   \n",
       "AACS      1.038974        0.518157  2.502786  0.207032  0.835985  1.000000   \n",
       "AAED1     9.829811        1.880914  0.951986  1.975780  0.048180  0.163028   \n",
       "AAGAB     2.013459        0.515767  1.758700  0.293266  0.769319  1.000000   \n",
       "AAK1      0.816817        0.059553  2.702106  0.022039  0.982417  1.000000   \n",
       "AAMDC     1.314319       -0.937528  2.182265 -0.429613  0.667478  1.000000   \n",
       "\n",
       "            group  \n",
       "A1BG      B_cells  \n",
       "A1BG-AS1  B_cells  \n",
       "A2M-AS1   B_cells  \n",
       "A4GALT    B_cells  \n",
       "AAAS      B_cells  \n",
       "AACS      B_cells  \n",
       "AAED1     B_cells  \n",
       "AAGAB     B_cells  \n",
       "AAK1      B_cells  \n",
       "AAMDC     B_cells  "
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>stat</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "      <th>group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>5.669955</td>\n",
       "      <td>0.403825</td>\n",
       "      <td>1.067884</td>\n",
       "      <td>0.378154</td>\n",
       "      <td>0.705316</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>B_cells</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <td>0.738919</td>\n",
       "      <td>2.278757</td>\n",
       "      <td>3.406331</td>\n",
       "      <td>0.668977</td>\n",
       "      <td>0.503510</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>B_cells</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M-AS1</th>\n",
       "      <td>0.157668</td>\n",
       "      <td>0.021039</td>\n",
       "      <td>5.332329</td>\n",
       "      <td>0.003946</td>\n",
       "      <td>0.996852</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>B_cells</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A4GALT</th>\n",
       "      <td>2.649744</td>\n",
       "      <td>4.131701</td>\n",
       "      <td>2.716025</td>\n",
       "      <td>1.521231</td>\n",
       "      <td>0.128202</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>B_cells</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAAS</th>\n",
       "      <td>1.998178</td>\n",
       "      <td>-0.768574</td>\n",
       "      <td>1.708237</td>\n",
       "      <td>-0.449923</td>\n",
       "      <td>0.652766</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>B_cells</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AACS</th>\n",
       "      <td>1.038974</td>\n",
       "      <td>0.518157</td>\n",
       "      <td>2.502786</td>\n",
       "      <td>0.207032</td>\n",
       "      <td>0.835985</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>B_cells</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAED1</th>\n",
       "      <td>9.829811</td>\n",
       "      <td>1.880914</td>\n",
       "      <td>0.951986</td>\n",
       "      <td>1.975780</td>\n",
       "      <td>0.048180</td>\n",
       "      <td>0.163028</td>\n",
       "      <td>B_cells</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAGAB</th>\n",
       "      <td>2.013459</td>\n",
       "      <td>0.515767</td>\n",
       "      <td>1.758700</td>\n",
       "      <td>0.293266</td>\n",
       "      <td>0.769319</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>B_cells</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAK1</th>\n",
       "      <td>0.816817</td>\n",
       "      <td>0.059553</td>\n",
       "      <td>2.702106</td>\n",
       "      <td>0.022039</td>\n",
       "      <td>0.982417</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>B_cells</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAMDC</th>\n",
       "      <td>1.314319</td>\n",
       "      <td>-0.937528</td>\n",
       "      <td>2.182265</td>\n",
       "      <td>-0.429613</td>\n",
       "      <td>0.667478</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>B_cells</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 5
  },
  {
   "cell_type": "markdown",
   "id": "291ffe1fc83cea75",
   "metadata": {},
   "source": [
    "## DEA consensus\n",
    "\n",
    "Additionally, the `do.tl.rank_genes_consensus` allow to perform both single-cell and pseudo-bulk DEA and generate a dataframe that summarises everything."
   ]
  },
  {
   "cell_type": "code",
   "id": "daf480662f1ff00c",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-10-22T14:37:39.278047Z",
     "start_time": "2025-10-22T14:36:53.082565Z"
    }
   },
   "source": [
    "df = do.tl.rank_genes_consensus(\n",
    "    adata,\n",
    "    ctrl_cond=\"healthy\",\n",
    "    disease_cond=\"disease\",\n",
    "    cluster_key=\"annotation\",\n",
    "    batch_key=\"batch\",\n",
    "    condition_key=\"condition\",\n",
    "    min_cells=30,\n",
    "    min_counts=10,\n",
    "    pseudobulk_approach=\"sum\",\n",
    "    technical_replicates=2,\n",
    "    get_results=True,\n",
    "    test_pseudobulk=\"edger\",\n",
    "    test=\"wilcoxon\",\n",
    ")"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-10-22 16:36:53,084 - Running wilcoxon\n",
      "2025-10-22 16:36:53,151 - Running DGEs for B_cells.\n",
      "2025-10-22 16:36:53,153 - Running wilcoxon test.\n",
      "2025-10-22 16:36:53,480 - Running DGEs for Monocytes.\n",
      "2025-10-22 16:36:53,482 - Running wilcoxon test.\n",
      "2025-10-22 16:36:53,565 - Running DGEs for NK.\n",
      "2025-10-22 16:36:53,566 - Running wilcoxon test.\n",
      "2025-10-22 16:36:53,677 - Running DGEs for T_cells.\n",
      "2025-10-22 16:36:53,680 - Running wilcoxon test.\n",
      "2025-10-22 16:36:54,161 - Running DGEs for pDC.\n",
      "2025-10-22 16:36:54,163 - Running wilcoxon test.\n",
      "2025-10-22 16:36:54,197 - Running edger\n",
      "2025-10-22 16:36:54,197 - Generating Pseudo-bulk data\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Pseudo-bulked groups: 100%|██████████| 10/10 [00:00<00:00, 33554.43it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-10-22 16:36:57,684 - Removed 5806 genes for having less than 10 total counts\n",
      "2025-10-22 16:36:57,699 - Run edgeR\n",
      "2025-10-22 16:36:57,703 - Running DEA for B_cells\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Reading AnnData in R\n",
      "Running edgeR Test\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-10-22 16:37:08,520 - Running DEA for Monocytes\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Generating DGE Table to pass to Python\n",
      "Reading AnnData in R\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-10-22 16:37:18,426 - Test could not be computed for Monocytes due to [Errno 2] No such file or directory: '/tmp/EdgeR_Test_eb8e58c43d00401db58be2acaae8a194/dge_Monocytes_edgeR.csv'\n",
      "2025-10-22 16:37:18,427 - Running DEA for NK\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Running edgeR Test\n",
      "Error in `contrasts<-`(`*tmp*`, value = contr.funs[1 + isOF[nn]]) : \n",
      "  contrasts can be applied only to factors with 2 or more levels\n",
      "Calls: fit_model ... model.matrix -> model.matrix.default -> contrasts<-\n",
      "Execution halted\n",
      "Reading AnnData in R\n",
      "Running edgeR Test\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-10-22 16:37:28,661 - Running DEA for T_cells\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Generating DGE Table to pass to Python\n",
      "Reading AnnData in R\n",
      "Running edgeR Test\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-10-22 16:37:38,776 - Generating consensus DataFrame\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Generating DGE Table to pass to Python\n"
     ]
    }
   ],
   "execution_count": 6
  },
  {
   "cell_type": "code",
   "id": "3632fbcd84d4abb7",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-10-22T14:37:49.180867Z",
     "start_time": "2025-10-22T14:37:49.168866Z"
    }
   },
   "source": [
    "df.head(10)"
   ],
   "outputs": [
    {
     "data": {
      "text/plain": [
       "   GeneName  wilcox_score     log2fc         pvals          padj  pts_group  \\\n",
       "0      JUND     10.587980   2.167520  3.388231e-26  1.254797e-22   0.982558   \n",
       "1      CD83     10.266931   4.102563  9.931388e-25  3.064992e-21   0.895349   \n",
       "2    RPS4Y1     10.197039  32.221443  2.044093e-24  5.407211e-21   0.813953   \n",
       "3       FOS     10.120420   4.805681  4.484900e-24  1.038086e-20   0.912791   \n",
       "4  HSP90AA1      9.908383   2.990328  3.827898e-23  7.875687e-20   0.959302   \n",
       "5      CREM      9.665054   5.431766  4.243946e-22  7.144105e-19   0.802326   \n",
       "6     YPEL5      9.301365   3.094231  1.386545e-20  1.974973e-17   0.889535   \n",
       "7      SRGN      9.120118   4.315721  7.504225e-20  8.684733e-17   0.802326   \n",
       "8    TUBB4B      8.940316   3.670844  3.880572e-19  3.992031e-16   0.784884   \n",
       "9      RGS1      8.825177   7.311319  1.092864e-18  9.233199e-16   0.697674   \n",
       "\n",
       "    pts_ref    group annotation  log2fc_edger  stat_edger    pval_edger  \\\n",
       "0  0.941176  disease    B_cells      2.150242  266.374184  1.508208e-09   \n",
       "1  0.176471  disease    B_cells      3.985247  289.033574  9.565300e-10   \n",
       "2  0.000000  disease    B_cells      9.754567  211.691989  1.603596e-08   \n",
       "3  0.308824  disease    B_cells      4.670979  681.658744  6.302140e-12   \n",
       "4  0.632353  disease    B_cells      3.064642  316.093902  5.622099e-10   \n",
       "5  0.058824  disease    B_cells      5.363856  299.596476  7.086365e-10   \n",
       "6  0.323529  disease    B_cells      2.976510  147.040320  4.389385e-08   \n",
       "7  0.161765  disease    B_cells      4.067877  206.063003  6.646860e-09   \n",
       "8  0.147059  disease    B_cells      3.703473  165.803423  2.310259e-08   \n",
       "9  0.029412  disease    B_cells      7.306141  567.203038  4.569613e-12   \n",
       "\n",
       "     padj_edger sc_signicant psc_signicant consensus_significant  \\\n",
       "0  5.454230e-07          Yes           Yes                   Yes   \n",
       "1  3.818415e-07          Yes           Yes                   Yes   \n",
       "2  3.189553e-06          Yes           Yes                   Yes   \n",
       "3  1.253496e-08          Yes           Yes                   Yes   \n",
       "4  3.621600e-07          Yes           Yes                   Yes   \n",
       "5  3.621600e-07          Yes           Yes                   Yes   \n",
       "6  6.073501e-06          Yes           Yes                   Yes   \n",
       "7  1.652575e-06          Yes           Yes                   Yes   \n",
       "8  4.177369e-06          Yes           Yes                   Yes   \n",
       "9  1.253496e-08          Yes           Yes                   Yes   \n",
       "\n",
       "   MeanExpr_disease  MeanExpr_healthy  \n",
       "0          4.278455          2.823328  \n",
       "1          2.662119          0.574226  \n",
       "2          1.793008          0.000000  \n",
       "3          3.868105          0.984030  \n",
       "4          3.364896          1.507360  \n",
       "5          2.367836          0.202228  \n",
       "6          2.494147          0.833389  \n",
       "7          2.626802          0.497278  \n",
       "8          1.940737          0.384084  \n",
       "9          2.666053          0.080899  "
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneName</th>\n",
       "      <th>wilcox_score</th>\n",
       "      <th>log2fc</th>\n",
       "      <th>pvals</th>\n",
       "      <th>padj</th>\n",
       "      <th>pts_group</th>\n",
       "      <th>pts_ref</th>\n",
       "      <th>group</th>\n",
       "      <th>annotation</th>\n",
       "      <th>log2fc_edger</th>\n",
       "      <th>stat_edger</th>\n",
       "      <th>pval_edger</th>\n",
       "      <th>padj_edger</th>\n",
       "      <th>sc_signicant</th>\n",
       "      <th>psc_signicant</th>\n",
       "      <th>consensus_significant</th>\n",
       "      <th>MeanExpr_disease</th>\n",
       "      <th>MeanExpr_healthy</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>JUND</td>\n",
       "      <td>10.587980</td>\n",
       "      <td>2.167520</td>\n",
       "      <td>3.388231e-26</td>\n",
       "      <td>1.254797e-22</td>\n",
       "      <td>0.982558</td>\n",
       "      <td>0.941176</td>\n",
       "      <td>disease</td>\n",
       "      <td>B_cells</td>\n",
       "      <td>2.150242</td>\n",
       "      <td>266.374184</td>\n",
       "      <td>1.508208e-09</td>\n",
       "      <td>5.454230e-07</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>4.278455</td>\n",
       "      <td>2.823328</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CD83</td>\n",
       "      <td>10.266931</td>\n",
       "      <td>4.102563</td>\n",
       "      <td>9.931388e-25</td>\n",
       "      <td>3.064992e-21</td>\n",
       "      <td>0.895349</td>\n",
       "      <td>0.176471</td>\n",
       "      <td>disease</td>\n",
       "      <td>B_cells</td>\n",
       "      <td>3.985247</td>\n",
       "      <td>289.033574</td>\n",
       "      <td>9.565300e-10</td>\n",
       "      <td>3.818415e-07</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>2.662119</td>\n",
       "      <td>0.574226</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>RPS4Y1</td>\n",
       "      <td>10.197039</td>\n",
       "      <td>32.221443</td>\n",
       "      <td>2.044093e-24</td>\n",
       "      <td>5.407211e-21</td>\n",
       "      <td>0.813953</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>disease</td>\n",
       "      <td>B_cells</td>\n",
       "      <td>9.754567</td>\n",
       "      <td>211.691989</td>\n",
       "      <td>1.603596e-08</td>\n",
       "      <td>3.189553e-06</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>1.793008</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>FOS</td>\n",
       "      <td>10.120420</td>\n",
       "      <td>4.805681</td>\n",
       "      <td>4.484900e-24</td>\n",
       "      <td>1.038086e-20</td>\n",
       "      <td>0.912791</td>\n",
       "      <td>0.308824</td>\n",
       "      <td>disease</td>\n",
       "      <td>B_cells</td>\n",
       "      <td>4.670979</td>\n",
       "      <td>681.658744</td>\n",
       "      <td>6.302140e-12</td>\n",
       "      <td>1.253496e-08</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>3.868105</td>\n",
       "      <td>0.984030</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HSP90AA1</td>\n",
       "      <td>9.908383</td>\n",
       "      <td>2.990328</td>\n",
       "      <td>3.827898e-23</td>\n",
       "      <td>7.875687e-20</td>\n",
       "      <td>0.959302</td>\n",
       "      <td>0.632353</td>\n",
       "      <td>disease</td>\n",
       "      <td>B_cells</td>\n",
       "      <td>3.064642</td>\n",
       "      <td>316.093902</td>\n",
       "      <td>5.622099e-10</td>\n",
       "      <td>3.621600e-07</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>3.364896</td>\n",
       "      <td>1.507360</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CREM</td>\n",
       "      <td>9.665054</td>\n",
       "      <td>5.431766</td>\n",
       "      <td>4.243946e-22</td>\n",
       "      <td>7.144105e-19</td>\n",
       "      <td>0.802326</td>\n",
       "      <td>0.058824</td>\n",
       "      <td>disease</td>\n",
       "      <td>B_cells</td>\n",
       "      <td>5.363856</td>\n",
       "      <td>299.596476</td>\n",
       "      <td>7.086365e-10</td>\n",
       "      <td>3.621600e-07</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>2.367836</td>\n",
       "      <td>0.202228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>YPEL5</td>\n",
       "      <td>9.301365</td>\n",
       "      <td>3.094231</td>\n",
       "      <td>1.386545e-20</td>\n",
       "      <td>1.974973e-17</td>\n",
       "      <td>0.889535</td>\n",
       "      <td>0.323529</td>\n",
       "      <td>disease</td>\n",
       "      <td>B_cells</td>\n",
       "      <td>2.976510</td>\n",
       "      <td>147.040320</td>\n",
       "      <td>4.389385e-08</td>\n",
       "      <td>6.073501e-06</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>2.494147</td>\n",
       "      <td>0.833389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>SRGN</td>\n",
       "      <td>9.120118</td>\n",
       "      <td>4.315721</td>\n",
       "      <td>7.504225e-20</td>\n",
       "      <td>8.684733e-17</td>\n",
       "      <td>0.802326</td>\n",
       "      <td>0.161765</td>\n",
       "      <td>disease</td>\n",
       "      <td>B_cells</td>\n",
       "      <td>4.067877</td>\n",
       "      <td>206.063003</td>\n",
       "      <td>6.646860e-09</td>\n",
       "      <td>1.652575e-06</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>2.626802</td>\n",
       "      <td>0.497278</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>TUBB4B</td>\n",
       "      <td>8.940316</td>\n",
       "      <td>3.670844</td>\n",
       "      <td>3.880572e-19</td>\n",
       "      <td>3.992031e-16</td>\n",
       "      <td>0.784884</td>\n",
       "      <td>0.147059</td>\n",
       "      <td>disease</td>\n",
       "      <td>B_cells</td>\n",
       "      <td>3.703473</td>\n",
       "      <td>165.803423</td>\n",
       "      <td>2.310259e-08</td>\n",
       "      <td>4.177369e-06</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>1.940737</td>\n",
       "      <td>0.384084</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>RGS1</td>\n",
       "      <td>8.825177</td>\n",
       "      <td>7.311319</td>\n",
       "      <td>1.092864e-18</td>\n",
       "      <td>9.233199e-16</td>\n",
       "      <td>0.697674</td>\n",
       "      <td>0.029412</td>\n",
       "      <td>disease</td>\n",
       "      <td>B_cells</td>\n",
       "      <td>7.306141</td>\n",
       "      <td>567.203038</td>\n",
       "      <td>4.569613e-12</td>\n",
       "      <td>1.253496e-08</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>2.666053</td>\n",
       "      <td>0.080899</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 7
  },
  {
   "cell_type": "code",
   "id": "b9f5175d889b478e",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-10-22T14:37:54.584834Z",
     "start_time": "2025-10-22T14:37:54.579643Z"
    }
   },
   "source": [
    "adata"
   ],
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 2783 × 18517\n",
       "    obs: 'batch', 'condition', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'total_counts_mt', 'log1p_total_counts_mt', 'pct_counts_mt', 'total_counts_ribo', 'log1p_total_counts_ribo', 'pct_counts_ribo', 'n_genes', 'n_counts', 'doublet_class', 'doublet_score', 'leiden', 'autoAnnot', 'celltypist_conf_score', 'annotation', 'annotation_recluster'\n",
       "    var: 'highly_variable', 'means', 'dispersions', 'dispersions_norm', 'highly_variable_nbatches', 'highly_variable_intersection'\n",
       "    uns: 'annotation_recluster_colors', 'hvg', 'leiden', 'log1p', 'neighbors', 'umap', 'rank_genes_pseudobulk', 'rank_genes_condition', 'rank_genes_consensus'\n",
       "    obsm: 'X_CCA', 'X_pca', 'X_umap'\n",
       "    layers: 'counts', 'logcounts'\n",
       "    obsp: 'connectivities', 'distances'"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 8
  },
  {
   "cell_type": "markdown",
   "id": "f9aa8de0fbe7594",
   "metadata": {},
   "source": "As we can appreciate, the results of the DEA will be saved in the uns attributed"
  },
  {
   "cell_type": "code",
   "id": "95fa2255680e56bc",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-10-22T14:38:04.177676Z",
     "start_time": "2025-10-22T14:38:02.156294Z"
    }
   },
   "source": [
    "\n",
    "session_info.show(na=False, cpu=True, excludes=[\"backports\"], std_lib=True, dependencies=True, html=True)"
   ],
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ],
      "text/html": [
       "<details>\n",
       "<summary>Click to view session information</summary>\n",
       "<pre>\n",
       "-----\n",
       "anndata             0.11.4\n",
       "dotools_py          0.0.1\n",
       "pandas              2.3.2\n",
       "session_info        v1.0.1\n",
       "sys                 3.11.13 (main, Jun  5 2025, 08:21:08) [Clang 14.0.6 ]\n",
       "-----\n",
       "</pre>\n",
       "<details>\n",
       "<summary>Click to view modules imported as dependencies</summary>\n",
       "<pre>\n",
       "Cython                      3.1.4\n",
       "PIL                         11.3.0\n",
       "absl                        2.3.1\n",
       "adjustText                  1.3.0\n",
       "appnope                     0.1.4\n",
       "argparse                    1.1\n",
       "arrow                       1.3.0\n",
       "attr                        25.3.0\n",
       "attrs                       25.3.0\n",
       "babel                       2.17.0\n",
       "certifi                     2025.08.03\n",
       "cffi                        2.0.0\n",
       "charset_normalizer          3.4.3\n",
       "cloudpickle                 3.1.1\n",
       "comm                        0.2.3\n",
       "coverage                    7.11.0\n",
       "csv                         1.0\n",
       "ctypes                      1.1.0\n",
       "cycler                      0.12.1\n",
       "cython                      3.1.4\n",
       "dask                        2024.11.2\n",
       "dateutil                    2.9.0.post0\n",
       "debugpy                     1.8.17\n",
       "decimal                     1.70\n",
       "decorator                   5.2.1\n",
       "defusedxml                  0.7.1\n",
       "deprecated                  1.2.18\n",
       "docrep                      0.3.2\n",
       "executing                   2.2.1\n",
       "formulaic                   1.2.0\n",
       "formulaic_contrasts         1.0.0\n",
       "fsspec                      2025.9.0\n",
       "geopandas                   1.1.1\n",
       "h5py                        3.14.0\n",
       "idna                        3.10\n",
       "igraph                      0.11.9\n",
       "interface_meta              1.3.0\n",
       "ipaddress                   1.0\n",
       "ipykernel                   6.30.1\n",
       "ipywidgets                  8.1.7\n",
       "jedi                        0.19.2\n",
       "jinja2                      3.1.6\n",
       "joblib                      1.5.2\n",
       "json                        2.0.9\n",
       "json5                       0.12.1\n",
       "jsonpointer                 3.0.0\n",
       "jsonschema                  4.25.1\n",
       "jupyter_events              0.12.0\n",
       "jupyter_server              2.17.0\n",
       "jupyterlab_server           2.27.3\n",
       "kiwisolver                  1.4.9\n",
       "lark                        1.2.2\n",
       "leidenalg                   0.10.2\n",
       "lightning                   2.5.5\n",
       "lightning_utilities         0.15.2\n",
       "llvmlite                    0.45.0\n",
       "logging                     0.5.1.2\n",
       "markupsafe                  3.0.2\n",
       "marshal                     4\n",
       "matplotlib                  3.10.6\n",
       "matplotlib_inline           0.1.7\n",
       "ml_collections              1.1.0\n",
       "mpmath                      1.3.0\n",
       "mudata                      0.3.2\n",
       "narwhals                    2.5.0\n",
       "natsort                     8.4.0\n",
       "nbformat                    5.10.4\n",
       "numba                       0.62.0\n",
       "numcodecs                   0.15.1\n",
       "numpy                       2.3.3\n",
       "opt_einsum                  3.4.0\n",
       "packaging                   25.0\n",
       "parso                       0.8.5\n",
       "patsy                       1.0.1\n",
       "platform                    1.0.8\n",
       "platformdirs                4.4.0\n",
       "polars                      1.33.1\n",
       "prompt_toolkit              3.0.52\n",
       "psutil                      7.1.0\n",
       "pure_eval                   0.2.3\n",
       "pyarrow                     21.0.0\n",
       "pycparser                   2.23\n",
       "pydeseq2                    0.5.2\n",
       "pygments                    2.19.2\n",
       "pyparsing                   3.2.4\n",
       "pyproj                      3.7.2\n",
       "pyro                        1.9.1\n",
       "pytz                        2025.2\n",
       "re                          2.2.1\n",
       "requests                    2.32.5\n",
       "rfc3339_validator           0.1.4\n",
       "rfc3986_validator           0.1.1\n",
       "scanpy                      1.11.3\n",
       "scipy                       1.15.3\n",
       "scvi                        1.4.0\n",
       "seaborn                     0.13.2\n",
       "shapely                     2.1.2\n",
       "six                         1.17.0\n",
       "sklearn                     1.7.2\n",
       "sniffio                     1.3.1\n",
       "socketserver                0.4\n",
       "sparse                      0.17.0\n",
       "sqlite3                     2.6.0\n",
       "stack_data                  0.6.3\n",
       "statsmodels                 0.14.5\n",
       "stdlib_list                 0.11.1\n",
       "sympy                       1.14.0\n",
       "tarfile                     0.9.0\n",
       "texttable                   1.7.0\n",
       "threadpoolctl               3.6.0\n",
       "tlz                         1.0.0\n",
       "toolz                       1.0.0\n",
       "torch                       2.8.0\n",
       "torchmetrics                1.8.2\n",
       "tornado                     6.5.2\n",
       "tqdm                        4.67.1\n",
       "traitlets                   5.14.3\n",
       "urllib3                     2.5.0\n",
       "wcwidth                     0.2.13\n",
       "websocket                   1.8.0\n",
       "wrapt                       1.17.3\n",
       "xarray                      2025.9.0\n",
       "yaml                        6.0.2\n",
       "zarr                        2.18.7\n",
       "zlib                        1.0\n",
       "zmq                         27.1.0\n",
       "</pre>\n",
       "</details> <!-- seems like this ends pre, so might as well be explicit -->\n",
       "<pre>\n",
       "-----\n",
       "IPython             9.5.0\n",
       "jupyter_client      8.6.3\n",
       "jupyter_core        5.8.1\n",
       "jupyterlab          4.4.7\n",
       "notebook            7.4.5\n",
       "-----\n",
       "Python 3.11.13 (main, Jun  5 2025, 08:21:08) [Clang 14.0.6 ]\n",
       "macOS-26.0.1-arm64-arm-64bit\n",
       "10 logical CPU cores, arm\n",
       "-----\n",
       "Session information updated at 2025-10-22 16:38\n",
       "</pre>\n",
       "</details>"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 9
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "37cd9d961bfb1b46",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
